---
title: "PanCan IRWG miRNA Correlative Analysis"
author: "James A. Eddy"
date: '`r lubridate::today()`'
output:
  html_notebook:
    code_folding: hide
    fig_width: 8
    toc: yes
    toc_float: yes
---

# Summary

Exploring associations of miRNA expression levels across tumor types in TCGA with suggested correlates defined by the Immune Response Working Group:

+ Overall Leukocyte fraction  
+ Individual Relative CIBERSORT fraction  
+ Mutation Load  
+ TCR,BCR Diversity  
+ Expression of  
    + IFNG,IFN-gamma  
    + PRF1,Perforin  
    + GZMA,Granzyme A  
    + PDCD1,PD-1  
    + CD274,PD-L1  
    + PDCD1LG2,PD-L2  
    + IL10,IL-10  
    + TGFB1,TGF-beta  
    + IDO1,IDO  
    + HLA-A  


```{r setup, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE, cache=FALSE}
# Synapse client
library(synapseClient)

# parallel computing
library(parallel)

# viz packages
library(ggthemes)
library(viridis)
library(ggbeeswarm)

# packages for general data munging, formatting
library(feather)
library(stringr)
library(forcats)
library(broom)
library(tidyverse)

my_theme_bw <- function() {
    theme_bw() +
        theme(axis.title = element_text(face = "bold", size = rel(0.9)),
              legend.title = element_text(face = "bold"),
              plot.title = element_text(face = "bold"))
}
ggplot2::theme_set(my_theme_bw())
scale_fill_discrete <- function(...) scale_fill_colorblind(...)
scale_fill_continuous <- function(...) scale_fill_viridis(...)
scale_colour_discrete <- function(...) scale_color_colorblind(...)
scale_colour_continuous <- function(...) scale_color_viridis(...)
```

-----

\  

# Prepare data

## Retrieve miRNA data from Synapse

Data are stored on Synapse in the folder `syn6171109`.

```{r, message=FALSE, warning=FALSE, include=FALSE}
if (!dir.exists("data")) {
    dir.create("data")
}
synapseLogin()
mirna_synfolder <- "syn6171109"
mirna_synfiles <- synapseQuery(
    sprintf('select * from file where parentId=="%s"', mirna_synfolder)
)

# download files and store data/paths in new data frame
mirna_files <- mirna_synfiles %>% 
    mutate(file_data = map(file.id, function(synid) {
        synGet(synid, downloadLocation = "../data/tcga/")
    }),
    file_path = map_chr(file_data, getFileLocation))
```

## Load miRNA sample data

Sample characteristics are stored in a tab-delimited text file (Synapse ID: `syn7222010`) and can be loaded with `read_tsv()`.

```{r}
# load sample data
mirna_sample_file <- mirna_files %>% 
    filter(file.id == "syn7222010") %>% 
    .[["file_path"]]
mirna_sample_df <- read_tsv(mirna_sample_file)
```

## Sample filtering

Sample Quality Annotations:
syn4551248 ("merged_sample_quality_annotations.tsv")

```{r}
sample_qual_file <- synGet("syn4551248", downloadLocation = "../data/tcga/")
sample_qual_df <- sample_qual_file %>% 
    getFileLocation() %>% 
    read_tsv()
```

remove samples based on `Do_not_use=True`, and remove cases with `AWG_excluded_because_of_pathology=True`

```{r}
# samples to exclude from all datasets
exclude_samples <- sample_qual_df %>% 
    mutate(AWG_excluded_because_of_pathology = parse_logical(AWG_excluded_because_of_pathology),
           Do_not_use = parse_logical(str_to_upper(Do_not_use))) %>% 
    filter(AWG_excluded_because_of_pathology | Do_not_use)
```

Remove samples from miRNA dataset.

```{r}
mirna_sample_df <- mirna_sample_df %>% 
    filter(!(id %in% exclude_samples$aliquot_barcode))
```

\  

## Load miRNA expression data

miRNA normalized, batch corrected expression values for all samples are stored as a matrix in a CSV file (Synapse ID: `syn7201053`) and can be loaded with `read_csv()`.

```{r warning=FALSE}
mirna_corr_file <- "../data/mirna_correlate_data.feather"
force_read <- FALSE
if (!file.exists(mirna_corr_file) | force_read) {
    # load normalized, batch-corrected expression data
    mirna_norm_file <- mirna_files %>% 
        filter(file.id == "syn7201053") %>% 
        .[["file_path"]]
    mirna_norm_df <- read_csv(mirna_norm_file, progress = FALSE)
    
    mirna_corr_df <- mirna_norm_df %>% 
        select(-one_of(exclude_samples$aliquot_barcode))
    
    write_feather(mirna_corr_df, mirna_corr_file)
} else {
    mirna_corr_df <- read_feather(mirna_corr_file)
}
```

\  

## Set up plotting colors

```{r}
tcga_colors <- tribble(
    ~Color, ~Disease,
    "#ED2891", "BRCA",
    "#B2509E", "GBM",
    "#D49DC7", "LGG",
    "#C1A72F", "ACC",
    "#E8C51D", "PCPG",
    "#F9ED32", "THCA",
    "#CACCDB", "CHOL",
    "#9EDDF9", "COAD",
    "#007EB5", "ESCA",
    "#104A7F", "LIHC",
    "#6E7BA2", "PAAD",
    "#DAF1FC", "READ",
    "#00AEEF", "STAD",
    "#F6B667", "CESC",
    "#D97D25", "OV",
    "#F89420", "UCEC",
    "#FBE3C7", "UCS",
    "#009444", "HNSC",
    "#97D1A9", "UVM",
    "#754C29", "LAML",
    "#CEAC8F", "THYM",
    "#3953A4", "DLBC",
    "#BBD642", "SKCM",
    "#00A99D", "SARC",
    "#542C88", "LUAD",
    "#A084BD", "LUSC",
    "#D3C3E0", "MESO",
    "#FAD2D9", "BLCA",
    "#EA7075", "KICH",
    "#ED1C24", "KIRC",
    "#F8AFB3", "KIRP",
    "#7E1918", "PRAD",
    "#BE1E2D", "TGCT"
)

mirna_sample_df <- mirna_sample_df %>% 
    mutate(Disease = factor(Disease, levels = tcga_colors$Disease))
```

-----

\  

# Explore miRNA data

## Sample characteristics

The table `mirna_sample_df` contains `r ncol(mirna_sample_df)` columns describing the `r nrow(mirna_sample_df)` samples in the data. The `Sample_Type` column corresponds to TCGA **sample type codes**, which are defined [here](https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes). The following sample types are included in the miRNA data:

```{r}
mirna_sample_df %>% 
    group_by(Sample_Type) %>% 
    tally()
```

\  

Based on the codes, this is the distribution of sample types:

```{r}
mirna_sample_df %>% 
    group_by(Sample_Type) %>% 
    tally() %>% 
    mutate(Definition = case_when(
        .$Sample_Type == 1 ~ "Primary Solid Tumor",
        .$Sample_Type == 2 ~ "Recurrent Solid Tumorr",
        .$Sample_Type == 3 ~ "Primary Blood Derived Cancer - Peripheral Blood",
        .$Sample_Type == 5 ~ "Additional - New Primary",
        .$Sample_Type == 6 ~ "Metastatic",
        .$Sample_Type == 7 ~ "Additional Metastatic",
        .$Sample_Type == 11 ~ "Solid Tissue Normal"
    ))
```

Note: only include "primary" tumor samples for now; also, add additional column to store vial ID (to ease mapping between data sets).

```{r}
mirna_sample_pt_df <- mirna_sample_df %>% 
    filter(Sample_Type %in% c(1, 3, 5)) %>% 
    mutate(vial_id = str_replace(id, "(\\-[:alnum:]+){3}$", ""))
```

\  

Additionally, the following disease types are included:

```{r}
mirna_sample_pt_df %>% 
    group_by(Disease) %>% 
    tally()
```

Note: exclude LAML, THYM, and DLBC from cell content correlations

\  

And technical variables:

```{r}
mirna_sample_pt_df %>% 
    group_by(Protocol, Platform) %>% 
    tally()
```

\  

## miRNA expression values

Expression values in the data are reportedly reads per million (RPM). I've randomly selected a few samples from each disease type, sample type, protocol, and platform to inspect the distribution of expression values across all `r nrow(mirna_norm_df)` miRNA genes.

```{r}
set.seed(0)

# randomly select 1-3 samples from each combination of characteristics
sample_sub_df <- mirna_sample_pt_df %>% 
    group_by(Disease, Sample_Type, Protocol, Platform) %>% 
    sample_n(3, replace = TRUE) %>% 
    ungroup() %>% 
    distinct()

# subset and melt the expression data
mirna_sub_df <- mirna_corr_df %>% 
    select(one_of(c("Genes", sample_sub_df$id))) %>%
    gather("sample", "expression", -Genes)
```

\  

Even with a shifted log (`log10(x + 1)`) transformation, expression values look to be more exponentially distributed than normal.

```{r}
mirna_sub_df %>% 
    left_join(mirna_sample_pt_df, by = c("sample" = "id")) %>% 
    ggplot(aes(x = log10(expression + 1))) +
    stat_density(aes(group = sample), geom = "line", position = "identity", 
                 alpha = 0.2)
```

\  

I can also look at the distribution of log-RPM values with boxplots:

```{r}
mirna_sub_df %>% 
    left_join(mirna_sample_pt_df, by = c("sample" = "id")) %>% 
    ggplot(aes(x = sample, y = log10(expression + 1))) +
    geom_boxplot(alpha = 0.5, outlier.size = 0.5) +
    theme(axis.text.x = element_blank())
```


```{r, warning=FALSE}
# convert expression df to matrix
mirna_corr_mat <- mirna_corr_df %>% 
    select(one_of(c("Genes", mirna_sample_pt_df$id))) %>% 
    column_to_rownames("Genes") %>% 
    as.matrix()
```
\  

## PCA

I used the `prcomp()` function on the transposed expression matrix (samples x genes, after transposing) to compute PCA data, which I can use to look for batch effects among samples.

```{r}
mirna_corr_pca <- mirna_corr_mat %>% 
    t() %>% 
    prcomp()
```

(I can use the `broom:tidy()` function to convert data in the `prcomp` object into data frames for `ggplot`.)

```{r, warning=FALSE}
pc_df <- mirna_corr_pca %>% 
    tidy("pcs")

mirna_corr_pca_df <- mirna_corr_pca %>% 
    tidy("samples") %>% 
    filter(PC <= 2) %>% 
    left_join(mirna_sample_pt_df, by = c("row" = "id"))
```

The plot below shows samples plotted as points along the first two principle components (PCs). Points are colored by disease type.

```{r}
pc1_label <- pc_df %>% 
    filter(PC == 1) %>% 
    transmute(label = sprintf("PC%s [%0.2f%%]", PC, 100*percent)) %>% 
    flatten_chr()

pc2_label <- pc_df %>% 
    filter(PC == 2) %>% 
    transmute(label = sprintf("PC%s [%0.2f%%]", PC, 100*percent)) %>% 
    flatten_chr()

mirna_corr_pca_df %>% 
    spread(PC, value) %>% 
    ggplot(aes(x = `1`, y = `2`)) +
    geom_point(aes(colour = Disease)) +
    xlab(pc1_label) +
    ylab(pc2_label) +
    scale_color_manual(values = tcga_colors$Color)
```

\  

Facetting by sample type...

```{r}
mirna_corr_pca_df %>% 
    spread(PC, value) %>% 
    ggplot(aes(x = `1`, y = `2`)) +
    geom_point(aes(colour = Disease)) +
    xlab(pc1_label) +
    ylab(pc2_label) +
    scale_color_manual(values = tcga_colors$Color) +
    facet_wrap(~ Sample_Type)
```

-----

\  

# Prepare correlate data

+ Overall Leukocyte fraction  
+ Individual Relative CIBERSORT fraction  
+ Mutation Load  
+ TCR,BCR Diversity  
+ Gene Expression

## Leukocyte fraction

### Retrieve/load data

Cellular Content: syn7994728
syn5808205 ("TCGA_all_leuk_estimate.masked.20170107.tsv")

```{r}
# file_data <- synGet("syn5808205", downloadLocation = "./")
leuk_frac_file <- "../data/tcga/TCGA_all_leuk_estimate.masked.20170107.tsv"
leuk_frac_df <- read_tsv(leuk_frac_file, col_names = FALSE) %>% 
    set_names(c("disease", "id", "leuk_frac"))
```

### Sample filtering

```{r}
leuk_frac_df <- leuk_frac_df %>%
    filter(!(id %in% exclude_samples$aliquot_barcode))
```


### Sample matching

Identify matched samples between miRNA and leukocyte fraction.

```{r}
leuk_frac_ids <- leuk_frac_df %>% 
    select(id) %>%
    mutate(vial_id = str_replace(id, "(\\-[:alnum:]+){3}$", "")) %>% 
    arrange()

mirna_ids <- mirna_sample_pt_df %>%
    select(id, vial_id) %>%
    arrange()

mirna_leuk_frac_shared_ids <- inner_join(mirna_ids, leuk_frac_ids, 
                                         by = "vial_id",
                                         suffix = c("_mirna", "_leuk_frac"))

# only keep samples with matched vial ID AND portion number
portion_id_minus_analyte_regex <- "([:alnum:]+\\-){4}[0-9]+"
# plate_id_only_regex <- "[:alnum:]+(?=([:alnum:]\\-[:alnum:]+){1}$)"
mirna_leuk_frac_shared_ids <- mirna_leuk_frac_shared_ids %>%
    filter((str_extract(id_mirna, portion_id_minus_analyte_regex)
            == str_extract(id_leuk_frac, portion_id_minus_analyte_regex)))
```

NOTE: several samples were assayed on multiple plates; average the leukocyte fraction across these before computing correlations with miRNA

### Correlate data formatting

```{r}
leuk_frac_corr_file <- "../data/leuk_frac_correlate_data.feather"
force_format <- FALSE
if (!file.exists(leuk_frac_corr_file) | force_format) {
    leuk_frac_corr_df <- mirna_leuk_frac_shared_ids %>% 
        left_join(leuk_frac_df, by = c("id_leuk_frac" = "id")) %>% 
        group_by(disease, vial_id) %>% 
        summarise(leuk_frac = mean(leuk_frac))
    
    write_feather(leuk_frac_corr_df, leuk_frac_corr_file)
} else {
    leuk_frac_corr_df <- read_feather(leuk_frac_corr_file)
}
```

### Disease-wise correlations

```{r warning=FALSE}
mirna_leuk_frac_corr_file <- "../data/mirna_leuk_frac_correlation.feather"
force_compute <- TRUE
if (!file.exists(mirna_leuk_frac_corr_file) | force_compute) {
    
    # skip immune cell cancers
    d_list <- mirna_sample_pt_df %>% 
        # filter((!Disease %in% c("LAML", "THYM", "DLBC"))) %>% 
        filter(Disease %in% leuk_frac_corr_df$disease) %>% 
        distinct(Disease) %>% 
        mutate(Disease = as.character(Disease)) %>% 
        .$Disease %>%
        set_names(.) %>% 
        as.list()
    
    corr_df_list <- mclapply(d_list, mc.cores = 4, function(d) {
        samples_d <- mirna_sample_pt_df %>%
            filter(Disease == d) %>%
            select(vial_id) %>%
            flatten_chr() %>%
            intersect(leuk_frac_corr_df$vial_id)

        source_df <- mirna_corr_df %>%
            set_names(str_replace(names(.), "(\\-[:alnum:]+){3}$", "")) %>%
            select(one_of(c("Genes", samples_d))) %>%
            dplyr::rename(mirna = Genes) %>%
            gather(sample, x, -mirna) %>% 
            filter(!is.na(x))
        
        target_df <- leuk_frac_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            dplyr::rename(sample = vial_id, x = leuk_frac) %>% 
            mutate(correlate = "leuk_frac")
        
        corr_df <- inner_join(source_df, target_df,
                              by = "sample",
                              suffix = c("_source", "_target")) %>%
            group_by(mirna, correlate) %>%
            do(tidy(cor.test(.$x_source, .$x_target, method = "spearman"))) %>% 
            ungroup()
        corr_df[["p.adjust"]] <- p.adjust(corr_df$p.value, method = "BH")
        return(corr_df)
    })
    
    mirna_leuk_frac_corr_df <- bind_rows(corr_df_list, .id = "disease") %>% 
        mutate(correlate_type = "leukocyte fraction")
    write_feather(mirna_leuk_frac_corr_df, mirna_leuk_frac_corr_file)
} else {
    mirna_leuk_frac_corr_df <- read_feather(mirna_leuk_frac_corr_file)
}
```

\  

## CIBERSORT fraction

### Retrieve/load data

Cellular Content: syn4991611
syn8024565 ("TCGA.cluster-by-CIBERSORT-relative.tsv")

or...?

syn7337221 ("TCGA.Kallisto.fullIDs.cibersort.relative.tsv")

```{r}
# ciber_frac_file <- "../data/tcga/TCGA.cluster-by-CIBERSORT-relative.tsv"
ciber_frac_file <- "../data/tcga/TCGA.Kallisto.fullIDs.cibersort.relative.tsv"

ciber_frac_df <- read_tsv(ciber_frac_file)
```

### Sample filtering

```{r}
ciber_frac_df <- ciber_frac_df %>%
    mutate(id = str_replace_all(SampleID, "\\.", "\\-")) %>% 
    filter(!(id %in% exclude_samples$aliquot_barcode))
```

### Sample matching

Identify matched samples between miRNA and CIBERSORT fraction.

```{r}
ciber_frac_ids <- ciber_frac_df %>% 
    select(id) %>%
    mutate(vial_id = str_replace(id, "(\\-[:alnum:]+){3}$", "")) %>% 
    arrange()

# only keep samples with matched vial ID AND portion number
mirna_ciber_frac_shared_ids <- inner_join(mirna_ids, ciber_frac_ids,
                                         by = "vial_id",
                                         suffix = c("_mirna", "_ciber_frac")) %>% 
    filter((str_extract(id_mirna, portion_id_minus_analyte_regex) 
            == str_extract(id_ciber_frac, portion_id_minus_analyte_regex)))
```

NOTE: several samples were assayed on multiple plates; average the CIBERSORT fraction across these before computing correlations with miRNA

### Correlate data formatting

```{r}
ciber_frac_corr_file <- "../data/ciber_frac_correlate_data.feather"
force_format <- FALSE
if (!file.exists(ciber_frac_corr_file) | force_format) {
    ciber_frac_corr_df <- mirna_ciber_frac_shared_ids %>%
        left_join(ciber_frac_df, by = c("id_ciber_frac" = "id")) %>% 
        select(-id_mirna, -id_ciber_frac, 
               -SampleID, -P.value, -Correlation, -RMSE) %>% 
        group_by(CancerType, vial_id) %>%
        summarise_each(funs(mean))
    
    write_feather(ciber_frac_corr_df, ciber_frac_corr_file)
} else {
    ciber_frac_corr_df <- read_feather(ciber_frac_corr_file)
}
```

### Disease-wise correlations

```{r warning=FALSE}
mirna_ciber_frac_corr_file <- "../data/mirna_ciber_frac_correlation.feather"
force_compute <- TRUE
if (!file.exists(mirna_ciber_frac_corr_file) | force_compute) {
    
    # skip immune cell cancers
    d_list <- mirna_sample_pt_df %>%
        filter((!Disease %in% c("LAML", "THCA", "DLBC")),
               Disease %in% ciber_frac_corr_df$CancerType) %>% 
        distinct(Disease) %>% 
        mutate(Disease = as.character(Disease)) %>% 
        .$Disease %>%
        set_names(.) %>% 
        as.list()
    
    corr_df_list <- mclapply(d_list, mc.cores = 4, function(d) {
        samples_d <- mirna_sample_pt_df %>%
            filter(Disease == d) %>%
            select(vial_id) %>%
            flatten_chr() %>%
            intersect(ciber_frac_corr_df$vial_id)

        source_df <- mirna_norm_df %>%
            set_names(str_replace(names(.), "(\\-[:alnum:]+){3}$", "")) %>%
            select(one_of(c("Genes", samples_d))) %>%
            dplyr::rename(mirna = Genes) %>%
            gather(sample, x, -mirna) %>% 
            filter(!is.na(x))

        target_df <- ciber_frac_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            select(-id_mirna, -id_ciber_frac, -CIBSERSORT.relative.CLUSTER,
                   -CancerType, -P.value, -Correlation,
                   -RMSE, -PatID, -samptype, -OS_Time, -OS_Status, 
                   -TotalLeukocyte, -gdTcells) %>% 
            gather(correlate, x, -vial_id) %>% 
            dplyr::rename(sample = vial_id)
        
        corr_df <- inner_join(source_df, target_df,
                              by = "sample",
                              suffix = c("_source", "_target")) %>%
            group_by(mirna, correlate) %>%
            do(tidy(cor.test(.$x_source, .$x_target, method = "spearman"))) %>% 
            ungroup()
        corr_df[["p.adjust"]] <- p.adjust(corr_df$p.value, method = "BH")
        return(corr_df)
    })
    
    mirna_ciber_frac_corr_df <- bind_rows(corr_df_list, .id = "disease") %>% 
        mutate(correlate_type = "CIBERSORT fraction")
    write_feather(mirna_ciber_frac_corr_df, mirna_ciber_frac_corr_file)
} else {
    mirna_ciber_frac_corr_df <- read_feather(mirna_ciber_frac_corr_file)
}
```

\  

## Mutation load

### Retrieve/load data

Mutation Load: syn5706827
syn7994728 ("mutation-load")

* exclude LAML?

```{r}
mut_load_file <- "../data/tcga/mutation-load-vial.txt"
mut_load_df <- read_tsv(mut_load_file)
```

### Sample filtering

Can't filter aliquots, as data only includes IDs specific to the level of vial

### Sample matching

Identify matched samples between miRNA and mutational load.

```{r}
mut_load_ids <- mut_load_df %>% 
    select(id = Tumor_Sample_ID_Vial) %>% 
    mutate(vial_id = id) %>%
    arrange()

mirna_mut_load_shared_ids <- inner_join(mirna_ids, mut_load_ids, 
                                        by = "vial_id",
                                        suffix = c("_mirna", "_mut_load"))
```

### Correlate data formatting

```{r}
mut_load_corr_file <- "../data/mut_load_correlate_data.feather"
force_format <- FALSE
if (!file.exists(mut_load_corr_file) | force_format) {
    mut_load_corr_df <- mirna_mut_load_shared_ids %>% 
        left_join(mut_load_df, by = c("id_mut_load" = "Tumor_Sample_ID_Vial"))
    
    write_feather(mut_load_corr_df, mut_load_corr_file)
} else {
    mut_load_corr_df <- read_feather(mut_load_corr_file)
}
```

### Disease-wise correlations

```{r warning=FALSE}
mirna_mut_load_corr_file <- "../data/mirna_mut_load_correlation.feather"
force_compute <- TRUE
if (!file.exists(mirna_mut_load_corr_file) | force_compute) {
    
    # skip immune cell cancers
    d_list <- mirna_sample_pt_df %>%
        filter(Disease %in% mut_load_corr_df$cohort) %>% 
        distinct(Disease) %>% 
        mutate(Disease = as.character(Disease)) %>% 
        .$Disease %>%
        set_names(.) %>% 
        as.list()
    
    corr_df_list <- mclapply(d_list, mc.cores = 4, function(d) {
        samples_d <- mirna_sample_pt_df %>%
            filter(Disease == d) %>%
            select(vial_id) %>%
            flatten_chr() %>%
            intersect(mut_load_corr_df$vial_id)

        source_df <- mirna_norm_df %>%
            set_names(str_replace(names(.), "(\\-[:alnum:]+){3}$", "")) %>%
            select(one_of(c("Genes", samples_d))) %>%
            dplyr::rename(mirna = Genes) %>%
            gather(sample, x, -mirna) %>% 
            filter(!is.na(x))

        target_df <- mut_load_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            select(-id_mirna, -id_mut_load, -cohort, -Patient_ID, 
                   -Tumor_Sample_ID) %>% 
            gather(correlate, x, -vial_id) %>% 
            dplyr::rename(sample = vial_id)
        
        corr_df <- inner_join(source_df, target_df,
                              by = "sample",
                              suffix = c("_source", "_target")) %>%
            group_by(mirna, correlate) %>%
            do(tidy(cor.test(.$x_source, .$x_target, method = "spearman"))) %>% 
            ungroup()
        corr_df[["p.adjust"]] <- p.adjust(corr_df$p.value, method = "BH")
        return(corr_df)
    })
    
    mirna_mut_load_corr_df <- bind_rows(corr_df_list, .id = "disease") %>% 
        mutate(correlate_type = "mutation load")
    write_feather(mirna_mut_load_corr_df, mirna_mut_load_corr_file)
} else {
    mirna_mut_load_corr_df <- read_feather(mirna_mut_load_corr_file)
}
```

\  

## TCR/BCR diversity

### Retrieve/load data

T Cell Receptor / Brown et al: syn5876488
syn7063422 ("mitcr_sampleStatistics_20160714.tsv")

```{r}
tcr_div_file <- "../data/tcga/mitcr_sampleStatistics_20160714.tsv"

tcr_div_df <- read_tsv(tcr_div_file)
```

### Sample filtering

```{r}
tcr_dv_df <- tcr_div_df %>%
    filter(!(AliquotBarcode %in% exclude_samples$aliquot_barcode))
```

### Sample matching

Identify matched samples between miRNA and CIBERSORT fraction.

```{r}
tcr_div_ids <- tcr_div_df %>% 
    select(id = AliquotBarcode, vial_id = SampleBarcode) %>% 
    arrange()

# only keep samples with matched vial ID AND portion number
mirna_tcr_div_shared_ids <- inner_join(mirna_ids, tcr_div_ids,
                                       by = "vial_id",
                                       suffix = c("_mirna", "_tcr_div")) %>% 
    filter((str_extract(id_mirna, portion_id_minus_analyte_regex) 
            == str_extract(id_tcr_div, portion_id_minus_analyte_regex))) %>% 
    filter((duplicated(vial_id) | duplicated(vial_id, fromLast = TRUE)))
```

NOTE: several samples were assayed on multiple plates; average the TCR diversity across these before computing correlations with miRNA

### Correlate data formatting

```{r}
# tcr_div_corr_file <- "../data/tcr_div_correlate_data.feather"
# force_format <- FALSE
# if (!file.exists(tcr_div_corr_file) | force_format) {
#     # tcr_div_corr_df <- 
#         
# mirna_tcr_div_shared_ids %>%
#         left_join(tcr_div_df, by = c("id_tcr_div" = "AliquotBarcode")) %>% 
#         # group_by(Study, vial_id) %>% 
#         # summarise(shannon = mean(shannon)) %>% 
#         # group_by(CancerType, vial_id) %>%
#         # summarise_each(funs(mean)) %>% 
#         I
#     
#     write_feather(tcr_div_corr_df, tcr_div_corr_file)
# } else {
#     tcr_div_corr_df <- read_feather(tcr_div_corr_file)
# }
```

## Gene expression

+ Expression of  
    + IFNG,IFN-gamma  
    + PRF1,Perforin  
    + GZMA,Granzyme A  
    + PDCD1,PD-1  
    + CD274,PD-L1  
    + PDCD1LG2,PD-L2  
    + IL10,IL-10  
    + TGFB1,TGF-beta  
    + IDO1,IDO  
    + HLA-A  

### Retrieve/load mRNA data

Batch effects normalized mRNA data: syn4976363
syn4976369 ("EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv")
syn4976366 (" EB++GeneExpAnnotation.tsv")

```{r include=FALSE}
mrna_synfolder <- "syn4976363"
mrna_synfiles <- synapseQuery(
    sprintf('select * from file where parentId=="%s"', mrna_synfolder)
)

# download files and store data/paths in new data frame
mrna_files <- mrna_synfiles %>% 
    mutate(file_data = map(file.id, function(synid) {
        synGet(synid, downloadLocation = "../data/tcga")
    }),
    file_path = map_chr(file_data, getFileLocation))
```

Sample characteristics are stored in a tab-delimited text file (Synapse ID: `syn4976366`) and can be loaded with `read_tsv()`.

```{r}
# load sample data
mrna_sample_file <- mrna_files %>% 
    filter(file.id == "syn4976366") %>% 
    .[["file_path"]]
mrna_sample_df <- read_tsv(mrna_sample_file)
```

### Sample filtering

Remove samples from mRNA dataset.

```{r}
mrna_sample_df <- mrna_sample_df %>% 
    filter(!(SampleID %in% exclude_samples$aliquot_barcode))
```

### Sample matching

Identify matched samples between miRNA and mRNA.

```{r}
mrna_ids <- mrna_sample_df %>% 
    select(SampleID) %>%
    mutate(vial_id = str_replace(SampleID, "(\\-[:alnum:]+){3}$", "")) %>% 
    arrange()

mirna_ids <- mirna_sample_pt_df %>%
    select(id, vial_id) %>%
    arrange()

mirna_mrna_shared_ids <- inner_join(mirna_ids, mrna_ids, by = "vial_id")

# only keep samples with matched vial ID AND portion number
mirna_mrna_shared_ids <- mirna_mrna_shared_ids %>%
    filter(str_extract(id, portion_id_minus_analyte_regex)
           == str_extract(SampleID, portion_id_minus_analyte_regex))
```

mRNA normalized, batch corrected expression values for all samples are stored as a matrix in a TSV file (Synapse ID: `syn4976369`) and can be loaded with `read_tsv()`.

### Correlate data formatting

List of genes accessed [here](https://docs.google.com/spreadsheets/d/1aqOXYsU1ubkbxIZI_5p8ZRootgOAT0KweMA3LvSZ7HY/edit#gid=0) and saved as a TSV at `data/Cancer Immunomodulators - TCGA PanImmune Group - Direct Relationship.tsv`:

```{r}
gene_correlate_file <- "../data/Cancer Immunomodulators - TCGA PanImmune Group - Direct Relationship.tsv"
gene_correlate_df <- read_tsv(gene_correlate_file)
```


```{r}
mrna_corr_file <- "../data/mrna_correlate_genes.feather"
force_format <- FALSE
if (!file.exists(mrna_corr_file) | force_format) {
    # load normalized, batch-corrected expression data
    mrna_norm_file <- mrna_files %>% 
        filter(file.id == "syn4976369") %>% 
        .[["file_path"]]
    mrna_norm_df <- read_tsv(mrna_norm_file, progress = FALSE)
    
    # gene_list <- c("IFNG", "PRF1", "GZMA", "PDCD1", "CD274", "PDCD1LG2", "IL10", 
    #                "TGFB1", "IDO1", "HLA-A")  
    mrna_corr_df <- mrna_norm_df %>% 
        separate(gene_id, c("gene_name", "gene_id"), sep = "\\|") %>% 
        filter((gene_id %in% gene_correlate_df$`Entrez ID`) 
               | (gene_name %in% gene_correlate_df$`HGNC Symbol`)) %>% 
        select(one_of(c("gene_name", "gene_id", 
                        mirna_mrna_shared_ids$SampleID)))
    write_feather(mrna_corr_df, mrna_corr_file)
} else {
    mrna_corr_df <- read_feather(mrna_corr_file)
}
```

### Disease-wise correlations

```{r message=FALSE, warning=FALSE}
mirna_mrna_corr_file <- "../data/mirna_mrna_correlation.feather"
force_compute <- TRUE
if (!file.exists(mirna_mrna_corr_file) | force_compute) {
    
    d_list <- mirna_sample_pt_df %>%
        distinct(Disease) %>% 
        mutate(Disease = as.character(Disease)) %>% 
        .$Disease %>%
        set_names(.) %>% 
        as.list()
    
    corr_df_list <- mclapply(d_list, mc.cores = 4, function(d) {
        samples_d <- mirna_sample_pt_df %>%
            filter(Disease == d) %>%
            select(vial_id) %>%
            flatten_chr() %>%
            intersect(str_replace(names(mrna_corr_df),
                                  "(\\-[:alnum:]+){3}$", ""))

        source_df <- mirna_norm_df %>%
            set_names(str_replace(names(.), "(\\-[:alnum:]+){3}$", "")) %>%
            select(one_of(c("Genes", samples_d))) %>%
            dplyr::rename(mirna = Genes) %>%
            gather(sample, x, -mirna) %>% 
            filter(!is.na(x))
        
        target_df <- mrna_corr_df %>%
            set_names(str_replace(names(.), "(\\-[:alnum:]+){3}$", "")) %>%
            select(one_of(c("gene_name", samples_d))) %>%
            dplyr::rename(correlate = gene_name) %>%
            gather(sample, x, -correlate) %>% 
            filter(!is.na(x))
        
        corr_df <- inner_join(source_df, target_df,
                              by = "sample",
                              suffix = c("_source", "_target")) %>%
            group_by(mirna, correlate) %>%
            do(tidy(cor.test(.$x_source, .$x_target, method = "spearman"))) %>% 
            ungroup()
        corr_df[["p.adjust"]] <- p.adjust(corr_df$p.value, method = "BH")
        return(corr_df)
    })
    
    mirna_mrna_corr_df <- bind_rows(corr_df_list, .id = "disease") %>% 
        mutate(correlate_type = "mRNA expression")
    write_feather(mirna_mrna_corr_df, mirna_mrna_corr_file)
} else {
    mirna_mrna_corr_df <- read_feather(mirna_mrna_corr_file)
}
```

### Correlation viz

```{r}
# mirna_mrna_corr_df %>% 
#     # filter(disease == "BRCA") %>% 
#     # group_by(mirna) %>% 
#     filter(p.adjust < 0.01, abs(estimate) > 0.5) %>%
#     select(mirna) %>% 
#     n_distinct()
#     # ungroup() %>% 
#     # left_join(gene_correlate_df, by = c("mrna" = "HGNC Symbol")) %>% 
#     # ggplot(aes(x = mrna, y = mirna)) +
#     # geom_tile(aes(fill = estimate)) +
#     # theme(axis.text.x = element_text(angle = 45, hjust = 1))
```



