---
title: "Exploring immunomodulator gene expression across TCGA immune subtypes"
output: 
  html_notebook:
      toc: true
      toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Summary

```{r setup, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE, cache=FALSE}
# Synapse client
library(synapseClient)

# parallel computing
library(parallel)

# viz packages
library(ggthemes)
library(viridis)
library(ggrepel)
library(ggbeeswarm)
library(ComplexHeatmap)
library(circlize)
library(patchwork)

# packages for general data munging, formatting
library(feather)
library(stringr)
library(forcats)
library(broom)
library(tidyverse)

# analysis packages
library(car)
library(vegan)

my_theme_bw <- function() {
    theme_bw() +
        theme(axis.title = element_text(face = "bold", size = rel(0.9)),
              legend.title = element_text(face = "bold"),
              plot.title = element_text(face = "bold"))
}
ggplot2::theme_set(my_theme_bw())
scale_fill_discrete <- function(...) scale_fill_colorblind(...)
scale_fill_continuous <- function(...) scale_fill_distiller(..., 
                                                            type = "div",
                                                            palette = "RdBu")
scale_colour_discrete <- function(...) scale_color_colorblind(...)

subtype_colors <- c("red", "yellow", "green", "cyan", "blue", "magenta")
```

# Prepare data

## Retrieve and load data

### IRWG feature matrix (Synapse)

```{r}
if (!dir.exists("../data")) {
    dir.create("../data")
}
synapseLogin()
```

*NOTE:* requires SFTP authentication

```{r}
irwg_fmx_file <- synapseClient::synGet("syn11187757", 
                                       downloadLocation = "../data/irwg/")
load(irwg_fmx_file@filePath)
irwg_fmx_df <- df
rm(df)
```


### TCGA gene expression data (Synapse)

Batch effects normalized mRNA data: syn4976363
syn4976369 ("EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv")
syn4976366 (" EB++GeneExpAnnotation.tsv")

```{r include=FALSE}
mrna_synfolder <- "syn4976363"
mrna_synfiles <- synapseClient::synapseQuery(
    sprintf('select * from file where parentId=="%s"', mrna_synfolder)
)

# download files and store data/paths in new data frame
mrna_files <- mrna_synfiles %>% 
    dplyr::mutate(file_data = map(file.id, function(synid) {
        synapseClient::synGet(synid, downloadLocation = "../data/tcga")
    }),
    file_path = purrr::map_chr(file_data, getFileLocation))
```


### Immunomodulator gene table (Google Sheets)

List of genes accessed [here](https://docs.google.com/spreadsheets/d/1aqOXYsU1ubkbxIZI_5p8ZRootgOAT0KweMA3LvSZ7HY/edit#gid=0) and saved as a TSV at `data/Cancer Immunomodulators - TCGA PanImmune Group - Direct Relationship.tsv`:

```{r}
im_url <- "https://docs.google.com/spreadsheets/d/1aqOXYsU1ubkbxIZI_5p8ZRootgOAT0KweMA3LvSZ7HY/edit#gid=0"
googlesheets::gs_auth()
im_gs <- googlesheets::gs_url(im_url)
```

```{r}
im_df <- googlesheets::gs_read(im_gs) %>% 
    dplyr::filter(!is.na(Gene))
```

## Format data for analysis

### Subset/format expression data

mRNA normalized, batch corrected expression values for all samples are stored as a matrix in a TSV file (Synapse ID: `syn4976369`) and can be loaded with `read_tsv()`. 

Only keep aliquots where `Use` is marked `1` in IRWG barcode mapping table.

Remove any genes with missing values in one or more samples.

```{r}
im_expr_file <- "../data/intermediate/immunomodulators_mrna-expr.feather"
force_format <- FALSE # loading full expression data is slow; avoid if possible
if (!file.exists(im_expr_file) | force_format) {
    # load normalized, batch-corrected expression data
    mrna_norm_file <- mrna_files %>% 
        dplyr::filter(file.id == "syn4976369") %>% 
        .[["file_path"]]
    mrna_norm_df <- readr::read_tsv(mrna_norm_file, progress = FALSE)
    
    im_expr_df <- mrna_norm_df %>%
        tidyr::separate(gene_id, c("gene_name", "gene_id"), sep = "\\|") %>% 
        dplyr::filter(gene_id %in% im_df$`Entrez ID` |
                          gene_name %in% im_df$`HGNC Symbol`) %>% 
        dplyr::select(-gene_id) %>% 
        tidyr::gather(AliquotBarcode, x, -gene_name) %>%
        dplyr::mutate(gene_name = stringr::str_c(gene_name, "RNASeq", 
                                                 sep = "_")) %>%
        tidyr::spread(gene_name, x) %>%
        dplyr::left_join(
            irwg_fmx_df %>% 
                select(-contains("RNASeq")), 
            ., 
            by = c("Representative_Expression_Matrix_AliquotBarcode" = 
                       "AliquotBarcode")) %>% 
        dplyr::filter(!is.na(Subtype_Immune_Model_Based))
    
    feather::write_feather(im_expr_df, im_expr_file)
} else {
    im_expr_df <- feather::read_feather(im_expr_file)
}
```


-----

# Analysis & Viz

## Expression distributions

```{r}
p1 <- im_expr_df %>% 
    dplyr::select(ParticipantBarcode, dplyr::contains("RNASeq")) %>% 
    tidyr::gather(gene_name, x, -ParticipantBarcode) %>%
    dplyr::group_by(gene_name) %>%
    # just check first 100 samples
    dplyr::slice(1:100) %>%
    dplyr::ungroup() %>% 
    ggplot2::ggplot(aes(x = log(x + 1))) +
    ggplot2::stat_density(ggplot2::aes(group = ParticipantBarcode), 
                          geom = "line", position = "identity",
                          alpha = 0.2)

p2 <- im_expr_df %>% 
    dplyr::select(ParticipantBarcode, Subtype_Immune_Model_Based,
                  dplyr::contains("RNASeq")) %>% 
    tidyr::gather(gene_name, x, 
                  -ParticipantBarcode, -Subtype_Immune_Model_Based) %>%
    dplyr::group_by(Subtype_Immune_Model_Based, gene_name) %>%
    # just check first 100 samples
    dplyr::slice(1:100) %>%
    dplyr::ungroup() %>% 
    ggplot2::ggplot(aes(x = log(x + 1))) +
    ggplot2::stat_density(ggplot2::aes(group = ParticipantBarcode), 
                          geom = "line", position = "identity",
                          alpha = 0.2) +
    ggplot2::facet_wrap(~ Subtype_Immune_Model_Based)

p1 + p2 + patchwork::plot_layout(ncol = 1)
```

## Expression heatmaps

### All immunomodulator genes

```{r}
im_logexpr_df <- im_expr_df %>% 
    # remove patients for which leukocyte fraction is missing
    dplyr::filter(!is.na(leukocyte_fraction)) %>%
    dplyr::select(ParticipantBarcode, Study, Subtype_Immune_Model_Based,
                  dplyr::contains("RNASeq")) %>%
    tidyr::gather(gene_name, x,
                  -ParticipantBarcode, -Study, -Subtype_Immune_Model_Based) %>%
    # remove genes with missing values
    dplyr::group_by(gene_name) %>%
    dplyr::filter(!any(is.na(x))) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(gene_name = stringr::str_replace(gene_name, "_RNASeq", ""),
                  log_x = log(x + 1)) %>%
    dplyr::left_join(im_df, by = c("gene_name" = "HGNC Symbol"))
```

```{r}
im_subtype_logexpr_df <- im_expr_df %>% 
    # remove patients for which leukocyte fraction is missing
    dplyr::filter(!is.na(leukocyte_fraction)) %>%
    dplyr::select(ParticipantBarcode, Study, Subtype_Immune_Model_Based,
                  dplyr::contains("RNASeq")) %>%
    tidyr::gather(gene_name, x,
                  -ParticipantBarcode, -Study, -Subtype_Immune_Model_Based) %>%
    # don't remove genes with missing values
    dplyr::mutate(gene_name = stringr::str_replace(gene_name, "_RNASeq", ""),
                  log_x = log(x + 1)) %>%
    dplyr::left_join(im_df, by = c("gene_name" = "HGNC Symbol")) %>%
    # filter missing values
    dplyr::filter(!is.na(log_x)) %>%
    dplyr::group_by(Subtype_Immune_Model_Based, gene_name) %>%
    dplyr::summarise(x_med = median(log_x),
                     n = length(log_x),
                     x_var = var(log_x)) %>%
    dplyr::ungroup() %>%
    dplyr::left_join(im_df, by = c("gene_name" = "HGNC Symbol")) %>%
    dplyr::mutate(`Immune Checkpoint` = dplyr::case_when(
        stringr::str_detect(`Immune Checkpoint`, "^I") ~ "Inhibitory",
        stringr::str_detect(`Immune Checkpoint`, "^S") ~ "Stimulatory",
        is.na(`Immune Checkpoint`) ~ "N/A",
        TRUE ~ `Immune Checkpoint`
    ))
readr::write_csv(im_subtype_logexpr_df,
                 "../data/intermediate/subtype_immunomodulator_expr.csv")
```


```{r}
im_subytype_logexpr_mat <- im_subtype_logexpr_df %>%
    dplyr::select(gene_name, Subtype_Immune_Model_Based, `Super Category`, 
                  x_med) %>%
    tidyr::spread(Subtype_Immune_Model_Based, x_med) %>%
    dplyr::mutate(mean_expr = sum(C1, C2, C3, C4, C5, C6) / 6) %>%
    dplyr::arrange(`Super Category`, mean_expr) %>%
    dplyr::select(-`Super Category`, -mean_expr) %>%
    tibble::remove_rownames() %>%
    tibble::column_to_rownames("gene_name") %>%
    as.matrix()

```

```{r, fig.asp=1.2, message=FALSE, warning=FALSE}
gene_annot_df <- tibble::tibble(label = rownames(im_subytype_logexpr_mat)) %>%
    dplyr::left_join(immunomod_df, by = c("label" = "HGNC Symbol")) %>%
    dplyr::mutate(`Immune Checkpoint` = dplyr::case_when(
        stringr::str_detect(`Immune Checkpoint`, "^I") ~ "Inhibitory",
        stringr::str_detect(`Immune Checkpoint`, "^S") ~ "Stimulatory",
        is.na(`Immune Checkpoint`) ~ "N/A",
        TRUE ~ `Immune Checkpoint`
    )) %>%
    tibble::column_to_rownames("label") %>%
    dplyr::select( `Immune Checkpoint`, `Super Category`) %>%
    as.data.frame()

category_gaps <- duplicated(gene_annot_df$`Super Category`) %>%
    magrittr::not() %>%
    which() %>%
    magrittr::subtract(1) %>%
    .[-1]

gene_annot_cols <- list(
    `Immune Checkpoint` = c(
        "Inhibitory" = "#333333",
        "Stimulatory" = "#E69F00",
        "N/A" = "#888888"
    ),
    `Super Category` = RColorBrewer::brewer.pal(7, "Set1") %>%
        set_names(unique(gene_annot_df$`Super Category`))
)
pheatmap::pheatmap(im_subytype_logexpr_mat,
                   color = rev(RColorBrewer::brewer.pal(11, "RdBu")),
                   annotation_row = gene_annot_df,
                   annotation_colors = gene_annot_cols,
                   annotation_names_row = FALSE,
                   cluster_cols = FALSE, cluster_rows = FALSE,
                   gaps_row = category_gaps,
                   fontsize = 6)
```


## PCA

### All immunomodulator genes

```{r, warning=FALSE}
# convert expression df to matrix
im_logexpr_mat <- im_logexpr_df %>%
    dplyr::select(ParticipantBarcode, log_x, gene_name) %>% 
    tidyr::spread(ParticipantBarcode, log_x) %>% 
    tibble::remove_rownames() %>% 
    tibble::column_to_rownames("gene_name") %>% 
    as.matrix()

im_logexpr_pca <- im_logexpr_mat %>% 
    t() %>% 
    prcomp()
```


#### Extract and format PC data

```{r, warning=FALSE}
pc_df <- im_logexpr_pca %>% 
    broom::tidy("pcs")

im_logexpr_pca_df <- im_logexpr_pca %>% 
    broom::tidy("samples") %>% 
    dplyr::filter(PC <= 2) %>% 
    dplyr::left_join(irwg_fmx_df, by = c("row" = "ParticipantBarcode"))
```


#### PCA bi-plots

```{r}
pc1_label <- pc_df %>% 
    dplyr::filter(PC == 1) %>% 
    dplyr::transmute(label = sprintf("PC%s [%0.2f%%]", PC, 100*percent)) %>% 
    purrr::flatten_chr()

pc2_label <- pc_df %>% 
    dplyr::filter(PC == 2) %>% 
    dplyr::transmute(label = sprintf("PC%s [%0.2f%%]", PC, 100*percent)) %>% 
    purrr::flatten_chr()

p1 <- im_logexpr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    ggplot2::ggplot(aes(x = `1`, y = `2`)) +
    ggplot2::geom_point(ggplot2::aes(fill = Subtype_Immune_Model_Based), 
                        size = 1.5, alpha = 0.8, shape = 21, colour = "black", 
                        stroke = 0.2) +
    ggplot2::xlab(pc1_label) +
    ggplot2::ylab(pc2_label) +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::theme(legend.position = "top") +
    ggplot2::coord_fixed()
p2 <- im_logexpr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    ggplot2::ggplot(aes(x = `1`, y = `2`)) +
    ggplot2::geom_point(ggplot2::aes(fill = leukocyte_fraction), 
                        size = 1.5, alpha = 0.8, shape = 21, colour = "black", 
                        stroke = 0.2) +
    ggplot2::xlab(pc1_label) +
    ggplot2::ylab(pc2_label) +
    ggplot2::theme(legend.position = "top") +
    ggplot2::coord_fixed()
p1 + p2
```

```{r}
im_logexpr_pca_df %>% 
    filter(PC == 1) %>% 
    ggplot(ggplot2::aes(x = leukocyte_fraction, y = value)) +
    ggplot2::geom_point(
        ggplot2::aes(fill = Subtype_Immune_Model_Based), 
        size = 1.5, alpha = 0.5,
        shape = 21, colour = "slategray", stroke = 0.2
    ) +
    ggplot2::geom_smooth(method = "lm") +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::guides(fill = FALSE) +
    ggplot2::ylab("PC1") +
    ggplot2::xlab("Leukocyte Fraction") +
    ggplot2::ggtitle("PC1 correlation with leukocyte fraction") +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8))
```


#### Bi-plots with loadings

```{r}
set.seed(0)
im_factor_df <- im_logexpr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    dplyr::select(row, leukocyte_fraction, Subtype_Immune_Model_Based) %>% 
    tibble::remove_rownames() %>% 
    tibble::column_to_rownames("row")

im_factor_scores <- vegan::envfit(im_logexpr_pca ~ 0 + ., 
                                  data = im_factor_df, perm = 1000)

im_arrow_df <- im_factor_scores$factors$centroids %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column("arrow") %>% 
    dplyr::mutate(
        factor = stringr::str_extract(arrow, 
                                      stringr::str_c(names(im_factor_df), 
                                                     collapse = "|")),
        arrow = stringr::str_replace(arrow, 
                                     stringr::str_c(names(im_factor_df), 
                                                    collapse = "|"), 
                                     "")
    ) %>% 
    dplyr::left_join(
        im_factor_scores$factors$r %>% 
            as.data.frame() %>% 
            tibble::rownames_to_column() %>%
            purrr::set_names(c("factor", "r2")),
        by = "factor"
    ) %>% 
    dplyr::left_join(
        im_factor_scores$factors$pvals %>% 
            as.data.frame() %>% 
            tibble::rownames_to_column() %>%
            purrr::set_names(c("factor", "pval")),
        by = "factor"
    ) %>% 
    dplyr::bind_rows(im_factor_scores$vectors$arrows %>% 
                         as.data.frame() %>% 
                         tibble::rownames_to_column("arrow") %>% 
                         dplyr::mutate(
                             factor = stringr::str_extract(
                                 arrow, 
                                 stringr::str_c(names(im_factor_df), 
                                                collapse = "|")
                             )
                         ) %>% 
                         dplyr::left_join(
                             im_factor_scores$vectors$r %>% 
                                 as.data.frame() %>% 
                                 tibble::rownames_to_column() %>%
                                 purrr::set_names(c("factor", "r2")),
                             by = "factor"
                         ) %>% 
                         dplyr::mutate(
                             pval = im_factor_scores$vectors$pvals
                         )
    ) %>% 
    dplyr::mutate(xlength = PC1 * sqrt(r2),
                  ylength = PC2 * sqrt(r2),
                  label = ifelse(pval < 0.05, arrow, ""))
    
```

```{r}
multiplier <- 1
im_logexpr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    ggplot2::ggplot(ggplot2::aes(x = `1`, y = `2`)) +
    ggplot2::stat_ellipse(geom = "polygon", alpha = 0.1) +
    ggplot2::coord_fixed() +
    ggplot2::geom_point(colour = "gray", size = 1.5, alpha = 0.5) +
    ggplot2::geom_segment(data = im_arrow_df,
                          ggplot2::aes(x = 0, xend = multiplier*xlength, 
                                       y = 0, yend = multiplier*ylength),
                          arrow = arrow(length = unit(.2, "cm"), 
                                        type = "closed"),
                          colour = "black") +
    ggrepel::geom_text_repel(data = im_arrow_df,
                             ggplot2::aes(x = multiplier*xlength, 
                                          y = multiplier*ylength, 
                                          label = label),
                             colour = "black", fontface = "bold", size = 4) +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::theme(legend.position = "top")

```

#### Per-subtype bi-plots

```{r}
bg_df <- im_logexpr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    dplyr::select(-Subtype_Immune_Model_Based)

p <- im_logexpr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    ggplot2::ggplot(aes(x = `1`, y = `2`)) +
    ggplot2::geom_point(data = bg_df, 
                        colour = "gray", size = 1, alpha = 0.5) +
    ggplot2::geom_point(ggplot2::aes(fill = leukocyte_fraction), 
                        size = 1, alpha = 0.8, shape = 21, colour = "black", 
                        stroke = 0.2) +
    ggplot2::scale_fill_viridis_c() +
    ggplot2::guides(fill = ggplot2::guide_colourbar(
        title = "Leukocyte Fraction", 
        title.position = "top")
    ) +
    ggplot2::xlab(pc1_label) +
    ggplot2::ylab(pc2_label) +
    ggplot2::coord_fixed() +
    ggplot2::facet_wrap(~ Subtype_Immune_Model_Based, ncol = 2) +
    my_theme_bw() +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.text = element_text(size = 7),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8),
                   strip.text.x = ggplot2::element_text(
                       face = "bold", 
                       angle = 0,
                       hjust = 0, 
                       size = 8,
                       margin = ggplot2::margin(l = 1, b = 2)),
                   strip.text.y = ggplot2::element_blank(),
                   strip.background = ggplot2::element_blank(),
                   legend.position = "bottom",
                   legend.direction = "horizontal") 
p
# ggplot2::ggsave("../figures/im_pca_lf.pdf", p,
#        width = 11, height = 12, units = "cm", dpi = 300)
```


## Clustering

```{r, fig.asp=1.2}
# skip_heatmap <- TRUE
# df1 <- tibble::tibble(label = colnames(immunomod_expr_mat)) %>%
#     dplyr::left_join(immunomod_expr_df, 
#                      by = c("label" = "ParticipantBarcode")) %>% 
#     select(Subtype_Immune_Model_Based, Study, leukocyte_fraction) %>% 
#     as.data.frame()
# ha <- HeatmapAnnotation(
#     df1, 
#     col = list(
#         Subtype_Immune_Model_Based = c(
#             "C1" = subtype_colors[1],
#             "C2" = subtype_colors[2],
#             "C3" = subtype_colors[3],
#             "C4" = subtype_colors[4],
#             "C5" = subtype_colors[5],
#             "C6" = subtype_colors[6]
#         ),
#         Study = tcga_colors %>% 
#             split(.$Disease) %>%
#             purrr::map("Color") %>% 
#             unlist(),
#         leukocyte_fraction = colorRamp2(seq(0, 1, length.out = 11), 
#                                         viridisLite::viridis(11))
#     ), annotation_legend_param = list(
#         Subtype_Immune_Model_Based = list(title = "Subtype"),
#         Study = list(nrow = 6),
#         leukocyte_fraction = list(title = "Leukocyte Fraction")
#     )
# )
# df2 <- tibble::tibble(label = rownames(immunomod_expr_mat)) %>% 
#     dplyr::left_join(immunomod_df, by = c("label" = "HGNC Symbol")) %>% 
#     dplyr::mutate(`Immune Checkpoint` = dplyr::case_when(
#         stringr::str_detect(`Immune Checkpoint`, "^I") ~ "Inhibitory",
#         stringr::str_detect(`Immune Checkpoint`, "^S") ~ "Stimulatory",
#         is.na(`Immune Checkpoint`) ~ "N/A",
#         TRUE ~ `Immune Checkpoint`
#     )) %>% 
#     dplyr::select(`Immune Checkpoint`) %>% 
#     as.data.frame()
# ra <- rowAnnotation(
#     df2,
#     col = list(
#         `Immune Checkpoint` = c(
#             "Inhibitory" = "#333333",
#             "Stimulatory" = "#E69F00",
#             "N/A" = "#888888"
#         )
#     )
# )
# ht <- ComplexHeatmap::Heatmap(immunomod_expr_mat, name = "log(Expression)",
#                         col = rev(RColorBrewer::brewer.pal(11, "RdBu")),
#                         show_column_names = FALSE,
#                         column_dend_height = unit(2, "cm"),
#                         row_names_gp = gpar(fontsize = 8),
#                         top_annotation = ha)
# if (!skip_heatmap) {
#     ComplexHeatmap::draw(ht + ra, annotation_legend_side = "bottom")    
# }
```



## Differential expression testing

### One-way ANOVA for gene expression ~ subtype

```{r, include=TRUE}
im_anova_df <- im_logexpr_df %>% 
    dplyr::group_by(gene_name) %>% 
    dplyr::do(
        broom::tidy(car::Anova(lm(log_x ~ 0 + Subtype_Immune_Model_Based, .),
                               white.adjust = TRUE))
    ) %>%
    dplyr::ungroup() %>% 
    dplyr::filter(term == "Subtype_Immune_Model_Based") %>%
    dplyr::arrange(dplyr::desc(statistic))

im_anova_df[["p.adjust"]] <- p.adjust(im_anova_df$p.value, 
                                             method = "BH")
im_anova_df %>%
    dplyr::select(gene_name, term, statistic, p.value, p.adjust)
```

```{r}
im_logexpr_df %>% 
    filter(gene_name == "HMGB1") %>% 
    lm(log_x ~ 0 + Subtype_Immune_Model_Based, 
       data = .) -> model
hist(residuals(model))
plot(model)
```

```{r}
im_logexpr_df %>% 
    dplyr::left_join(irwg_fmx_df %>% 
                         select(ParticipantBarcode, leukocyte_fraction), 
                     by = "ParticipantBarcode") %>% 
    filter(gene_name == "HMGB1") %>% 
    lm(log_x ~ 0 + Subtype_Immune_Model_Based + leukocyte_fraction, 
       data = .) -> model
hist(residuals(model))
plot(model)
```

### Two-way ANOVA for gene expression ~ subtype + leukocyte fraction

```{r, include=TRUE}
im_lf_anova_df <- im_logexpr_df %>% 
    dplyr::left_join(irwg_fmx_df %>% 
                         select(ParticipantBarcode, leukocyte_fraction), 
                     by = "ParticipantBarcode") %>% 
    dplyr::group_by(gene_name) %>% 
    dplyr::do(
        broom::tidy(car::Anova(lm(log_x ~ 0 + Subtype_Immune_Model_Based + leukocyte_fraction, .),
                               white.adjust = TRUE))
    ) %>%
    dplyr::ungroup() %>% 
    dplyr::filter(term == "Subtype_Immune_Model_Based") %>%
    dplyr::arrange(dplyr::desc(statistic))

im_lf_anova_df[["p.adjust"]] <- p.adjust(im_lf_anova_df$p.value, 
                                         method = "BH")
im_lf_anova_df %>%
    dplyr::select(gene_name, term, statistic, p.value, p.adjust)
```

### Kruskal-Wallis

```{r}
im_kw_df <- im_logexpr_df %>% 
    dplyr::left_join(irwg_fmx_df %>% 
                         dplyr::select(ParticipantBarcode, leukocyte_fraction), 
                     by = "ParticipantBarcode") %>% 
    dplyr::group_by(gene_name) %>%    
    dplyr::do(
        broom::tidy(kruskal.test(log_x ~ 0 + Subtype_Immune_Model_Based, .))
    ) %>%
    dplyr::ungroup() %>% 
    dplyr::arrange(desc(statistic))

im_kw_df[["p.adjust"]] <- p.adjust(im_kw_df$p.value, method = "BH")
im_kw_df %>% 
    select(gene_name, statistic, p.value, p.adjust)
```


### Top genes from Kruskal-Wallis

Picking a few of the top genes (based on test statistic... as p-value isn't very useful), I can see how their distributions are changing across clusters. There certainly seems to be a reasonable difference, but I don't know how interesting this is.

```{r}
top_genes <- im_kw_df %>% 
    dplyr::slice(1:2) %>% 
    .[["gene_name"]]
p <- im_logexpr_df %>% 
    dplyr::filter(gene_name %in% top_genes) %>%
    dplyr::mutate(gene_name = factor(gene_name, levels = top_genes)) %>% 
    ggplot2::ggplot(aes(y = log_x, x = Subtype_Immune_Model_Based)) +
    ggbeeswarm::geom_quasirandom(
        ggplot2::aes(fill = Subtype_Immune_Model_Based, 
                     colour = Subtype_Immune_Model_Based), 
        size = 1.5, alpha = 0.5,
        shape = 21, colour = "slategray", stroke = 0.5
    ) +
    ggplot2::geom_boxplot(
        ggplot2::aes(colour = Subtype_Immune_Model_Based), 
        alpha = 0.6, outlier.size = 0, outlier.alpha = 0, size = 0.2
    ) +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::scale_colour_manual(values = subtype_colors) +
    ggplot2::guides(colour = FALSE, fill = FALSE) +
    ggplot2::ylab("log10(x)") +
    ggplot2::xlab("") +
    ggplot2::facet_wrap(~ gene_name, ncol = 1) +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.text.x = ggplot2::element_text(size = 7),
                   axis.text.y = ggplot2::element_text(size = 7),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8),
                   strip.text.x = ggplot2::element_text(
                       face = "bold", angle = 0, hjust = 0, 
                       size = 8, margin = ggplot2::margin(l = 2, b = 2)
                   ),
                   strip.text.y = ggplot2::element_blank(),
                   strip.background = ggplot2::element_blank(),
                   panel.spacing.y = ggplot2::unit(4, "points"),
                   panel.spacing.x = ggplot2::unit(2, "points"))
p
# ggsave("../figures/im_subtype_expr_dists.pdf", p,
#        width = 4.5, height = 5, units = "cm", dpi = 300)
```

```{r}
p <- im_logexpr_df %>% 
    dplyr::left_join(irwg_fmx_df %>% 
                         dplyr::select(ParticipantBarcode, leukocyte_fraction), 
                     by = "ParticipantBarcode") %>% 
    dplyr::filter(gene_name %in% top_genes) %>%
    dplyr::mutate(gene_name = factor(gene_name, levels = top_genes)) %>% 
    ggplot2::ggplot(ggplot2::aes(y = log_x, x = leukocyte_fraction)) +
    ggplot2::geom_point(
        ggplot2::aes(fill = Subtype_Immune_Model_Based), 
        size = 1.5, alpha = 0.5,
        shape = 21, colour = "slategray", stroke = 0.5
    ) +
    ggplot2::geom_smooth(method = "lm") +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::guides(fill = FALSE) +
    ggplot2::ylab("log(Expression)") +
    ggplot2::xlab("Leukocyte Fraction") +
    ggplot2::facet_wrap(~ gene_name) +
    my_theme_bw() +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.text = ggplot2::element_text(size = 7),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8),
                   strip.text.x = element_text(
                       face = "bold", angle = 0, hjust = 0, 
                       size = 8, margin = ggplot2::margin(l = 2, b = 2)),
                   strip.text.y = ggplot2::element_blank(),
                   strip.background = ggplot2::element_blank())
p

# ggsave("../figures/im_lf_corr.pdf", p,
#        width = 6, height = 5, units = "cm", dpi = 300, scale = 1)
```

```{r}
im_lfresid_df <- im_logexpr_df %>% 
    dplyr::left_join(irwg_fmx_df %>% 
                         dplyr::select(ParticipantBarcode, leukocyte_fraction), 
                     by = "ParticipantBarcode") %>% 
    tidyr::nest(-gene_name) %>% 
    dplyr::mutate(
        fit = purrr::map(data, ~ lm(log_x ~ leukocyte_fraction, data = .)),
        results = purrr::map(fit, broom::augment)
    ) %>% 
    tidyr::unnest(data, results)
```

```{r}
p <- im_lfresid_df %>% 
    dplyr::filter(gene_name %in% top_genes) %>%
    dplyr::mutate(gene_name = factor(gene_name, levels = top_genes)) %>% 
    ggplot2::ggplot(aes(y = .resid, x = Subtype_Immune_Model_Based)) +
    ggbeeswarm::geom_quasirandom(
        ggplot2::aes(fill = Subtype_Immune_Model_Based), 
        size = 1.5, alpha = 0.5,
        shape = 21, colour = "slategray", stroke = 0.5
    ) +
    ggplot2::geom_boxplot(
        ggplot2::aes(colour = Subtype_Immune_Model_Based), 
        alpha = 0.6, outlier.size = 0, outlier.alpha = 0
    ) +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::scale_colour_manual(values = subtype_colors) +
    ggplot2::guides(colour = FALSE, fill = FALSE) +
    ggplot2::ylab("LF Residual Expression") +
    ggplot2::xlab("Subtype") +
    ggplot2::facet_wrap(~ gene_name, ncol = 1) +
    my_theme_bw() +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8),
                   axis.text = ggplot2::element_text(size = 7),
                   strip.text.x = ggplot2::element_text(
                       face = "bold", angle = 0, hjust = 0, 
                       size = 8, margin = ggplot2::margin(l = 2, b = 2)),
                   strip.text.y = ggplot2::element_blank(),
                   strip.background = ggplot2::element_blank())
p

# ggsave("../figures/im_lf_subtype_expr_dists.pdf", p,
#        width = 6, height = 6, units = "cm", dpi = 300, scale = 1)
```

```{r}
im_lfresid_kw_df <- im_lfresid_df %>% 
    dplyr::left_join(irwg_fmx_df %>% 
                         dplyr::select(ParticipantBarcode, leukocyte_fraction), 
                     by = "ParticipantBarcode") %>% 
    dplyr::group_by(gene_name) %>%    
    dplyr::do(
        broom::tidy(kruskal.test(.resid ~ 0 + Subtype_Immune_Model_Based, .))
    ) %>%
    dplyr::ungroup() %>% 
    dplyr::arrange(dplyr::desc(statistic))

im_lfresid_kw_df[["p.adjust"]] <- p.adjust(im_lfresid_kw_df$p.value, 
                                           method = "BH")
im_lfresid_kw_df %>% 
    select(gene_name, statistic, p.value, p.adjust)
```

