---
title: "PanCan IRWG miRNA Correlative Analysis"
author: "James A. Eddy"
date: '`r lubridate::today()`'
output:
  html_notebook:
    code_folding: hide
    fig_width: 8
    toc: yes
    toc_float: yes
---

# Summary

Exploring associations of miRNA expression levels across tumor types in TCGA with suggested correlates defined by the Immune Response Working Group:

+ Overall Leukocyte fraction  
+ Individual Relative CIBERSORT fraction  
+ Mutation Load  
+ TCR,BCR Diversity  
+ Expression of  
    + IFNG,IFN-gamma  
    + PRF1,Perforin  
    + GZMA,Granzyme A  
    + PDCD1,PD-1  
    + CD274,PD-L1  
    + PDCD1LG2,PD-L2  
    + IL10,IL-10  
    + TGFB1,TGF-beta  
    + IDO1,IDO  
    + HLA-A  


```{r setup, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE, cache=FALSE}
# Synapse client
library(synapseClient)

# viz packages
library(ggthemes)
library(viridis)
library(ggbeeswarm)

# packages for general data munging, formatting
library(feather)
library(stringr)
library(forcats)
library(broom)
library(tidyverse)

my_theme_bw <- function() {
    theme_bw() +
        theme(axis.title = element_text(face = "bold", size = rel(0.9)),
              legend.title = element_text(face = "bold"),
              plot.title = element_text(face = "bold"))
}
ggplot2::theme_set(my_theme_bw())
scale_fill_discrete <- function(...) scale_fill_colorblind(...)
scale_fill_continuous <- function(...) scale_fill_viridis(...)
scale_colour_discrete <- function(...) scale_color_colorblind(...)
scale_colour_continuous <- function(...) scale_color_viridis(...)
```

-----

\  

# Prepare data

## Retrieve miRNA data from Synapse

Data are stored on Synapse in the folder `syn6171109`.

```{r, message=FALSE, warning=FALSE}
if (!dir.exists("data")) {
    dir.create("data")
}
synapseLogin()
mirna_synfolder <- "syn6171109"
mirna_synfiles <- synapseQuery(
    sprintf('select * from file where parentId=="%s"', mirna_synfolder)
)

# download files and store data/paths in new data frame
mirna_files <- mirna_synfiles %>% 
    mutate(file_data = map(file.id, function(synid) {
        synGet(synid, downloadLocation = "data/")
    }),
    file_path = map_chr(file_data, getFileLocation))
```

## Load miRNA data

Sample characteristics are stored in a tab-delimited text file (Synapse ID: `syn7222010`) and can be loaded with `read_tsv()`.

```{r}
# load sample data
mirna_sample_file <- mirna_files %>% 
    filter(file.id == "syn7222010") %>% 
    .[["file_path"]]
mirna_sample_df <- read_tsv(mirna_sample_file)
```

miRNA normalized, batch corrected expression values for all samples are stored as a matrix in a CSV file (Synapse ID: `syn7201053`) and can be loaded with `read_csv()`.

```{r}
# load normalized, batch-corrected expression data
mirna_norm_file <- mirna_files %>% 
    filter(file.id == "syn7201053") %>% 
    .[["file_path"]]
mirna_norm_df <- read_csv(mirna_norm_file, progress = FALSE)
```

```{r}
tcga_colors <- tribble(
    ~Color, ~Disease,
    "#ED2891", "BRCA",
    "#B2509E", "GBM",
    "#D49DC7", "LGG",
    "#C1A72F", "ACC",
    "#E8C51D", "PCPG",
    "#F9ED32", "THCA",
    "#CACCDB", "CHOL",
    "#9EDDF9", "COAD",
    "#007EB5", "ESCA",
    "#104A7F", "LIHC",
    "#6E7BA2", "PAAD",
    "#DAF1FC", "READ",
    "#00AEEF", "STAD",
    "#F6B667", "CESC",
    "#D97D25", "OV",
    "#F89420", "UCEC",
    "#FBE3C7", "UCS",
    "#009444", "HNSC",
    "#97D1A9", "UVM",
    "#754C29", "LAML",
    "#CEAC8F", "THYM",
    "#3953A4", "DLBC",
    "#BBD642", "SKCM",
    "#00A99D", "SARC",
    "#542C88", "LUAD",
    "#A084BD", "LUSC",
    "#D3C3E0", "MESO",
    "#FAD2D9", "BLCA",
    "#EA7075", "KICH",
    "#ED1C24", "KIRC",
    "#F8AFB3", "KIRP",
    "#7E1918", "PRAD",
    "#BE1E2D", "TGCT"
)

mirna_sample_df <- mirna_sample_df %>% 
    mutate(Disease = factor(Disease, levels = tcga_colors$Disease))
```

-----

\  

# Explore miRNA data

## Sample characteristics

The table `mirna_sample_df` contains `r ncol(mirna_sample_df)` columns describing the `r nrow(mirna_sample_df)` samples in the data. The `Sample_Type` column corresponds to TCGA **sample type codes**, which are defined [here](https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes). The following sample types are included in the miRNA data:

```{r}
mirna_sample_df %>% 
    group_by(Sample_Type) %>% 
    tally()
```

\  

Based on the codes, this is the distribution of sample types:

```{r}
mirna_sample_df %>% 
    group_by(Sample_Type) %>% 
    tally() %>% 
    mutate(Definition = case_when(
        .$Sample_Type == 1 ~ "Primary Solid Tumor",
        .$Sample_Type == 2 ~ "Recurrent Solid Tumorr",
        .$Sample_Type == 3 ~ "Primary Blood Derived Cancer - Peripheral Blood",
        .$Sample_Type == 5 ~ "Additional - New Primary",
        .$Sample_Type == 6 ~ "Metastatic",
        .$Sample_Type == 7 ~ "Additional Metastatic",
        .$Sample_Type == 11 ~ "Solid Tissue Normal"
    ))
```

Note: only include "primary" tumor samples for now; also, add additional column to store tumor ID (to ease mapping between data sets).

```{r}
mirna_sample_pt_df <- mirna_sample_df %>% 
    filter(Sample_Type %in% c(1, 3, 5)) %>% 
    mutate(tid = str_replace(id, "(\\-[:alnum:]+){2}$", ""))
```

\  

Additionally, the following disease types are included:

```{r}
mirna_sample_pt_df %>% 
    group_by(Disease) %>% 
    tally()
```

Note: exclude LAML, THCA, and DLBC from cell content correlations

\  

And technical variables:

```{r}
mirna_sample_pt_df %>% 
    group_by(Protocol, Platform) %>% 
    tally()
```

\  

## miRNA expression values

Expression values in the data are reportedly reads per million (RPM). I've randomly selected a few samples from each disease type, sample type, protocol, and platform to inspect the distribution of expression values across all `r nrow(mirna_norm_df)` miRNA genes.

```{r}
set.seed(0)

# randomly select 1-3 samples from each combination of characteristics
sample_sub_df <- mirna_sample_pt_df %>% 
    group_by(Disease, Sample_Type, Protocol, Platform) %>% 
    sample_n(3, replace = TRUE) %>% 
    ungroup() %>% 
    distinct()

# subset and melt the expression data
norm_sub_df <- mirna_norm_df %>% 
    select(one_of(c("Genes", sample_sub_df$id))) %>%
    gather("sample", "expression", -Genes)
```

\  

Even with a shifted log (`log10(x + 1)`) transformation, expression values look to be more exponentially distributed than normal.

```{r}
norm_sub_df %>% 
    left_join(mirna_sample_pt_df, by = c("sample" = "id")) %>% 
    ggplot(aes(x = log10(expression + 1))) +
    stat_density(aes(group = sample), geom = "line", position = "identity", 
                 alpha = 0.2)
```

\  

I can also look at the distribution of log-RPM values with boxplots:

```{r}
norm_sub_df %>% 
    left_join(mirna_sample_pt_df, by = c("sample" = "id")) %>% 
    ggplot(aes(x = sample, y = log10(expression + 1))) +
    geom_boxplot(alpha = 0.5, outlier.size = 0.5) +
    theme(axis.text.x = element_blank())
```


```{r, warning=FALSE}
# convert expression df to matrix
mirna_norm_mat <- mirna_norm_df %>% 
    select(one_of(c("Genes", mirna_sample_pt_df$id))) %>% 
    column_to_rownames("Genes") %>% 
    as.matrix()
```
\  

## PCA

I used the `prcomp()` function on the transposed expression matrix (samples x genes, after transposing) to compute PCA data, which I can use to look for batch effects among samples.

```{r}
mirna_norm_pca <- mirna_norm_mat %>% 
    t() %>% 
    prcomp()
```

(I can use the `broom:tidy()` function to convert data in the `prcomp` object into data frames for `ggplot`.)

```{r, warning=FALSE}
pc_df <- mirna_norm_pca %>% 
    tidy("pcs")

mirna_norm_pca_df <- mirna_norm_pca %>% 
    tidy("samples") %>% 
    filter(PC <= 2) %>% 
    left_join(mirna_sample_pt_df, by = c("row" = "id"))
```

The plot below shows samples plotted as points along the first two principle components (PCs). Points are colored by disease type.

```{r}
pc1_label <- pc_df %>% 
    filter(PC == 1) %>% 
    transmute(label = sprintf("PC%s [%0.2f%%]", PC, 100*percent)) %>% 
    flatten_chr()

pc2_label <- pc_df %>% 
    filter(PC == 2) %>% 
    transmute(label = sprintf("PC%s [%0.2f%%]", PC, 100*percent)) %>% 
    flatten_chr()

mirna_norm_pca_df %>% 
    spread(PC, value) %>% 
    ggplot(aes(x = `1`, y = `2`)) +
    geom_point(aes(colour = Disease)) +
    xlab(pc1_label) +
    ylab(pc2_label) +
    scale_color_manual(values = tcga_colors$Color)
```

\  

Facetting by sample type...

```{r}
mirna_norm_pca_df %>% 
    spread(PC, value) %>% 
    ggplot(aes(x = `1`, y = `2`)) +
    geom_point(aes(colour = Disease)) +
    xlab(pc1_label) +
    ylab(pc2_label) +
    scale_color_manual(values = tcga_colors$Color) +
    facet_wrap(~ Sample_Type)
```

-----

\  

# Prepare correlate data

+ Overall Leukocyte fraction  
+ Individual Relative CIBERSORT fraction  
+ Mutation Load  
+ TCR,BCR Diversity  
+ Gene Expression

## Leukocyte fraction

Cellular Content: syn7994728
syn5808205 ("TCGA_all_leuk_estimate.masked.20170107.tsv")

```{r}
# file_data <- synGet("syn5808205", downloadLocation = "./")
```

## CIBERSORT fraction

Cellular Content: syn7994728
syn8024565 ("TCGA.cluster-by-CIBERSORT-relative.tsv")

## Mutation load

Mutation Load: syn5706827
syn5706827 ("mutation-load")

## TCR/BCR diversity

## Gene expression

+ Expression of  
    + IFNG,IFN-gamma  
    + PRF1,Perforin  
    + GZMA,Granzyme A  
    + PDCD1,PD-1  
    + CD274,PD-L1  
    + PDCD1LG2,PD-L2  
    + IL10,IL-10  
    + TGFB1,TGF-beta  
    + IDO1,IDO  
    + HLA-A  

Batch effects normalized mRNA data: syn4976363
syn4976369 ("EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv")
syn4976366 (" EB++GeneExpAnnotation.tsv")

```{r}
mrna_synfolder <- "syn4976363"
mrna_synfiles <- synapseQuery(
    sprintf('select * from file where parentId=="%s"', mrna_synfolder)
)

# download files and store data/paths in new data frame
mrna_files <- mrna_synfiles %>% 
    mutate(file_data = map(file.id, function(synid) {
        synGet(synid, downloadLocation = "data/")
    }),
    file_path = map_chr(file_data, getFileLocation))
```

Sample characteristics are stored in a tab-delimited text file (Synapse ID: `syn4976366`) and can be loaded with `read_tsv()`.

```{r}
# load sample data
mrna_sample_file <- mrna_files %>% 
    filter(file.id == "syn4976366") %>% 
    .[["file_path"]]
mrna_sample_df <- read_tsv(mrna_sample_file)
```

Identify matched samples between miRNA and mRNA.

```{r}
mrna_ids <- mrna_sample_df %>% 
    select(SampleID) %>%
    mutate(tid = str_replace(SampleID, "(\\-[:alnum:]+){2}$", "")) %>% 
    arrange()

mirna_ids <- mirna_sample_pt_df %>%
    select(id, tid) %>%
    arrange()

tumor_ids <- inner_join(mirna_ids, mrna_ids, by = "tid")
```

mRNA normalized, batch corrected expression values for all samples are stored as a matrix in a TSV file (Synapse ID: `syn4976369`) and can be loaded with `read_tsv()`.

List of genes accessed [here](https://docs.google.com/spreadsheets/d/1aqOXYsU1ubkbxIZI_5p8ZRootgOAT0KweMA3LvSZ7HY/edit#gid=0) and saved as a TSV at `data/Cancer Immunomodulators - TCGA PanImmune Group - Direct Relationship.tsv`:

```{r}
gene_correlate_file <- "data/Cancer Immunomodulators - TCGA PanImmune Group - Direct Relationship.tsv"
gene_correlate_df <- read_tsv(gene_correlate_file)
```

Check if any genes from the table are missing from expression data...

```{r}
# check for discrepancies in gene name/ID
# mrna_norm_genes %>%
#     filter((gene_id %in% gene_correlate_df$`Entrez ID`) |
#            (gene_name %in% gene_correlate_df$`HGNC Symbol`)) %>%
#     left_join(gene_correlate_df %>%
#                   select(gene_name = `HGNC Symbol`,
#                          gene_id = `Entrez ID`),
#               by = "gene_name", suffix = c("_syn", "_cp")) %>%
#     left_join(gene_correlate_df %>%
#                   select(gene_name = `HGNC Symbol`,
#                          gene_id = `Entrez ID`) %>%
#                   mutate(gene_id = as.character(gene_id)),
#               by = c("gene_id_syn" =  "gene_id"), 
#               suffix = c("_syn", "_cp")) %>%
#     filter((gene_name_syn != gene_name_cp) | (gene_id_syn != gene_id_cp))
```


```{r}
mrna_corr_file <- "data/mrna_correlate_genes.feather"
if (!file.exists(mrna_corr_file)) {
    # load normalized, batch-corrected expression data
    mrna_norm_file <- mrna_files %>% 
        filter(file.id == "syn4976369") %>% 
        .[["file_path"]]
    mrna_norm_df <- read_tsv(mrna_norm_file, progress = FALSE)
    
    # gene_list <- c("IFNG", "PRF1", "GZMA", "PDCD1", "CD274", "PDCD1LG2", "IL10", 
    #                "TGFB1", "IDO1", "HLA-A")  
    mrna_corr_df <- mrna_norm_df %>% 
        separate(gene_id, c("gene_name", "gene_id"), sep = "\\|") %>% 
        filter((gene_id %in% gene_correlate_df$`Entrez ID`) 
               | (gene_name %in% gene_correlate_df$`HGNC Symbol`)) %>% 
        select(one_of(c("gene_name", "gene_id", tumor_ids$SampleID)))
    write_feather(mrna_corr_df, mrna_corr_file)
} else {
    mrna_corr_df <- read_feather(mrna_corr_file)
}
```

### Disease-wise correlations

```{r}
d_list <- mirna_sample_pt_df %>%
    distinct(Disease) %>% 
    mutate(Disease = as.character(Disease)) %>% 
    .$Disease %>%
    set_names(.) %>% 
    as.list()

corr_df_list <- map(d_list, function(d) {
    print(d)
    tumors_d <- mirna_sample_pt_df %>%
        filter(Disease == d) %>%
        select(tid) %>%
        flatten_chr() %>%
        intersect(str_replace(names(mrna_corr_df),
                              "(\\-[:alnum:]+){2}$", ""))
    print(length(tumors_d))
    source_df <- mirna_norm_df %>%
        set_names(str_replace(names(.), "(\\-[:alnum:]+){2}$", "")) %>%
        select(one_of(c("Genes", tumors_d))) %>%
        dplyr::rename(mirna = Genes) %>%
        gather(tumor, x, -mirna)

    target_df <- mrna_corr_df %>%
        set_names(str_replace(names(.), "(\\-[:alnum:]+){2}$", "")) %>%
        select(one_of(c("gene_name", tumors_d))) %>%
        dplyr::rename(mrna = gene_name) %>%
        gather(tumor, x, -mrna)

    corr_df <- inner_join(source_df, target_df,
                          by = "tumor",
                          suffix = c("_source", "_target")) %>%
        group_by(mirna, mrna) %>%
        do(tidy(cor.test(.$x_source, .$x_target, method = "spearman")))
    corr_df[["p.adjust"]] <- p.adjust(corr_df$p.value, method = "BH")
    return(corr_df)
})
```

```{r}
corr_df_list[[1]] %>% 
    group_by(mirna) %>% 
    filter(any(p.adjust < 0.001)) %>%
    ungroup() %>% 
    ggplot(aes(x = mrna, y = mirna)) +
    geom_tile(aes(fill = estimate))

sum(corr_df_list[[1]]$estimate == corr_df_list[[2]]$estimate, na.rm = TRUE)
```


