---
title: "R Notebook"
output: html_notebook
---

# Summary

```{r setup, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE, cache=FALSE}
# Synapse client
library(synapseClient)

# parallel computing
library(parallel)

# viz packages
library(ggthemes)
library(viridis)
library(ggrepel)
library(ggbeeswarm)
library(ComplexHeatmap)
library(circlize)
library(patchwork)

# packages for general data munging, formatting
library(feather)
library(stringr)
library(forcats)
library(broom)
library(tidyverse)

# analysis packages
library(car)
library(vegan)

my_theme_bw <- function() {
    theme_bw() +
        theme(axis.title = element_text(face = "bold", size = rel(0.9)),
              legend.title = element_text(face = "bold"),
              plot.title = element_text(face = "bold"))
}
ggplot2::theme_set(my_theme_bw())
scale_fill_discrete <- function(...) scale_fill_colorblind(...)
scale_fill_continuous <- function(...) scale_fill_distiller(..., 
                                                            type = "div",
                                                            palette = "RdBu")
scale_colour_discrete <- function(...) scale_color_colorblind(...)
# scale_colour_continuous <- function(...) scale_color_viridis(...)

subtype_colors <- c("red", "yellow", "green", "cyan", "blue", "magenta")

tcga_colors <- tribble(
    ~Color, ~Disease,
    "#ED2891", "BRCA",
    "#B2509E", "GBM",
    "#D49DC7", "LGG",
    "#C1A72F", "ACC",
    "#E8C51D", "PCPG",
    "#F9ED32", "THCA",
    "#104A7F", "CHOL",
    "#9EDDF9", "COAD",
    "#007EB5", "ESCA",
    "#CACCDB", "LIHC",
    "#6E7BA2", "PAAD",
    "#DAF1FC", "READ",
    "#00AEEF", "STAD",
    "#F6B667", "CESC",
    "#D97D25", "OV",
    "#FBE3C7", "UCEC",
    "#F89420", "UCS",
    "#97D1A9", "HNSC",
    "#009444", "UVM",
    "#754C29", "LAML",
    "#CEAC8F", "THYM",
    "#3953A4", "DLBC",
    "#BBD642", "SKCM",
    "#00A99D", "SARC",
    "#D3C3E0", "LUAD",
    "#A084BD", "LUSC",
    "#542C88", "MESO",
    "#FAD2D9", "BLCA",
    "#ED1C24", "KICH",
    "#F8AFB3", "KIRC",
    "#EA7075", "KIRP",
    "#7E1918", "PRAD",
    "#BE1E2D", "TGCT"
)
```

# Prepare data

## Retrieve and load data

### IRWG feature matrix (Synapse)

```{r}
if (!dir.exists("../data")) {
    dir.create("../data")
}
synapseLogin()
```

*NOTE:* requires SFTP authentication

```{r}
irwg_fmx_file <- synapseClient::synGet("syn11187757", 
                                       downloadLocation = "../data/irwg/")
load(irwg_fmx_file@filePath)
irwg_fmx_df <- df
rm(df)
```


### TCGA gene expression data (Synapse)

Batch effects normalized mRNA data: syn4976363
syn4976369 ("EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv")
syn4976366 (" EB++GeneExpAnnotation.tsv")

```{r include=FALSE}
mrna_synfolder <- "syn4976363"
mrna_synfiles <- synapseClient::synapseQuery(
    sprintf('select * from file where parentId=="%s"', mrna_synfolder)
)

# download files and store data/paths in new data frame
mrna_files <- mrna_synfiles %>% 
    dplyr::mutate(file_data = map(file.id, function(synid) {
        synapseClient::synGet(synid, downloadLocation = "../data/tcga")
    }),
    file_path = purrr::map_chr(file_data, getFileLocation))
```


### Immunomodulator gene table (Google Sheets)

List of genes accessed [here](https://docs.google.com/spreadsheets/d/1aqOXYsU1ubkbxIZI_5p8ZRootgOAT0KweMA3LvSZ7HY/edit#gid=0) and saved as a TSV at `data/Cancer Immunomodulators - TCGA PanImmune Group - Direct Relationship.tsv`:

```{r}
immunomod_url <- "https://docs.google.com/spreadsheets/d/1aqOXYsU1ubkbxIZI_5p8ZRootgOAT0KweMA3LvSZ7HY/edit#gid=0"
googlesheets::gs_auth()
immunomod_gs <- googlesheets::gs_url(immunomod_url)
```

```{r}
immunomod_df <- googlesheets::gs_read(immunomod_gs) %>% 
    dplyr::filter(!is.na(Gene))
```

## Format data for analysis

### Subset/format expression data

mRNA normalized, batch corrected expression values for all samples are stored as a matrix in a TSV file (Synapse ID: `syn4976369`) and can be loaded with `read_tsv()`. 

Only keep aliquots where `Use` is marked `1` in IRWG barcode mapping table.

Remove any genes with missing values in one or more samples.

```{r}
immunomod_expr_file <- "../data/intermediate/immunomodulators_mrna-expr.feather"
force_format <- FALSE # loading full expression data is slow; avoid if possible
if (!file.exists(immunomod_expr_file) | force_format) {
    # load normalized, batch-corrected expression data
    mrna_norm_file <- mrna_files %>% 
        dplyr::filter(file.id == "syn4976369") %>% 
        .[["file_path"]]
    mrna_norm_df <- readr::read_tsv(mrna_norm_file, progress = FALSE)
    
    immunomod_expr_df <- mrna_norm_df %>%
        tidyr::separate(gene_id, c("gene_name", "gene_id"), sep = "\\|") %>% 
        dplyr::filter(gene_id %in% immunomod_df$`Entrez ID` |
                          gene_name %in% immunomod_df$`HGNC Symbol`) %>% 
        dplyr::select(one_of(c("gene_name",
                               irwg_barcodemap_df$AliquotBarcode))) %>%
        tidyr::gather(AliquotBarcode, x, -gene_name) %>%
        dplyr::mutate(gene_name = str_c(gene_name, "RNASeq", sep = "_")) %>%
        dplyr::group_by(gene_name) %>% 
        dplyr::filter(!any(is.na(x))) %>%
        dplyr::ungroup() %>% 
        tidyr::spread(gene_name, x) %>%
        dplyr::left_join(irwg_fmx_df %>% 
                             select(-contains("RNASeq")), 
                         ., 
                         by = c("Representative_Expression_Matrix_AliquotBarcode" = "AliquotBarcode")) %>% 
        dplyr::filter_at(dplyr::vars(dplyr::contains("RNASeq")), 
                         dplyr::all_vars(!is.na(.)))
    
    feather::write_feather(immunomod_expr_df, immunomod_expr_file)
} else {
    immunomod_expr_df <- feather::read_feather(immunomod_expr_file)
}
```


### Remove patients with missing data

Remove patients for which leukocyte fraction is missing.

```{r}
immunomod_expr_df <- immunomod_expr_df %>% 
    dplyr::filter(!is.na(leukocyte_fraction)) %>% 
    dplyr::filter(!is.na(Subtype_Immune_Model_Based))
```

-----

# Analysis & Viz

## Expression distributions

```{r}
p1 <- immunomod_expr_df %>% 
    dplyr::select(ParticipantBarcode, dplyr::contains("RNASeq")) %>% 
    tidyr::gather(gene_name, x, -ParticipantBarcode) %>%
    group_by(gene_name) %>%
    # just check first 100 samples
    slice(1:100) %>%
    ggplot(aes(x = log(x + 1))) +
    stat_density(aes(group = ParticipantBarcode), 
                 geom = "line", position = "identity",
                 alpha = 0.2)
p2 <- immunomod_expr_df %>% 
    dplyr::select(ParticipantBarcode, Subtype_Immune_Model_Based,
                  dplyr::contains("RNASeq")) %>% 
    tidyr::gather(gene_name, x, 
                  -ParticipantBarcode, -Subtype_Immune_Model_Based) %>%
    group_by(Subtype_Immune_Model_Based, gene_name) %>%
    # just check first 100 samples
    slice(1:100) %>%
    ggplot(aes(x = log(x + 1))) +
    stat_density(aes(group = ParticipantBarcode), 
                 geom = "line", position = "identity",
                 alpha = 0.2) +
    facet_wrap(~ Subtype_Immune_Model_Based)
p1 + p2 + patchwork::plot_layout(ncol = 1)
```

```{r}
frontline_immunomod_genes <- c("BTLA", "C10orf54", "CD274", "CTLA4", "HAVCR2", 
                               "LAG3", "PDCD1", "PDCD1LG2", "TIGIT")
```

## Expression heatmaps

### All immunomodulator genes

```{r}
immunomod_zscore_df <- immunomod_expr_df %>% 
    dplyr::select(ParticipantBarcode, Study, Subtype_Immune_Model_Based,
                  dplyr::contains("RNASeq")) %>% 
    tidyr::gather(gene_name, x, 
                  -ParticipantBarcode, -Study, -Subtype_Immune_Model_Based) %>% 
    dplyr::mutate(gene_name = stringr::str_replace(gene_name, "_RNASeq", ""),
                  log_x = log(x + 1)) %>% 
    dplyr::left_join(immunomod_df, by = c("gene_name" = "HGNC Symbol")) %>% 
    dplyr::group_by(gene_name) %>% 
    dplyr::mutate(z_score = (log_x - mean(log_x)) / sd(log_x),
                  x_perc = log_x / max(log_x)) %>% 
    dplyr::ungroup()  
```

```{r, fig.height=8, fig.width=6}
immunomod_zscore_df %>% 
    dplyr::group_by(Subtype_Immune_Model_Based, gene_name) %>% 
    dplyr::mutate(x_med = median(log_x)) %>% 
    dplyr::ungroup() %>% 
    ggplot2::ggplot(ggplot2::aes(x = Subtype_Immune_Model_Based, 
                                 y = gene_name)) +
    ggplot2::geom_tile(aes(fill = x_med)) +
    ggplot2::scale_fill_distiller(type = "div", palette = "RdBu") +
    ggplot2::facet_grid(`Gene Family` ~ Subtype_Immune_Model_Based, 
                        space = "free", scales = "free", drop = TRUE) +
    ggplot2::theme(axis.text.x = ggplot2::element_blank(),
          axis.ticks.x = ggplot2::element_blank(),
          strip.text.x = ggplot2::element_text(face = "bold", angle = 0, 
                                               hjust = 0),
          strip.text.y = ggplot2::element_text(angle = 0, face = "bold", 
                                               hjust = 0),
          strip.background = ggplot2::element_blank(),
          panel.spacing.y = ggplot2::unit(0.5, "points"))
```

### Front-line immunomodulator genes

```{r}
immunomod_sub_zscore_df <- immunomod_expr_df %>% 
    dplyr::select(ParticipantBarcode, Study, Subtype_Immune_Model_Based,
                  dplyr::contains("RNASeq")) %>% 
    tidyr::gather(gene_name, x, 
                  -ParticipantBarcode, -Study, -Subtype_Immune_Model_Based) %>% 
    dplyr::mutate(gene_name = stringr::str_replace(gene_name, "_RNASeq", ""),
                  log_x = log(x + 1)) %>% 
    dplyr::filter(gene_name %in% frontline_immunomod_genes) %>% 
    dplyr::left_join(immunomod_df, by = c("gene_name" = "HGNC Symbol")) %>% 
    dplyr::group_by(gene_name) %>% 
    dplyr::mutate(z_score = (log_x - mean(log_x)) / sd(log_x),
                  x_perc = log_x / max(log_x)) %>% 
    dplyr::ungroup()  
```

```{r, include=FALSE}
immunomod_sub_zscore_df %>% 
    dplyr::group_by(Subtype_Immune_Model_Based, gene_name) %>% 
    dplyr::mutate(x_med = median(log_x)) %>% 
    dplyr::ungroup() %>% 
    ggplot2::ggplot(ggplot2::aes(x = Subtype_Immune_Model_Based, 
                                 y = gene_name)) +
    ggplot2::geom_tile(aes(fill = x_med)) +
    ggplot2::scale_fill_distiller(type = "div", palette = "RdBu") +
    ggplot2::facet_grid(`Gene Family` ~ Subtype_Immune_Model_Based, 
                        space = "free", scales = "free", drop = TRUE) +
    ggplot2::theme(axis.text.x = ggplot2::element_blank(),
          axis.ticks.x = ggplot2::element_blank(),
          strip.text.x = ggplot2::element_text(face = "bold", angle = 0, 
                                               hjust = 0),
          strip.text.y = ggplot2::element_text(angle = 0, face = "bold", 
                                               hjust = 0),
          strip.background = ggplot2::element_blank(),
          panel.spacing.y = ggplot2::unit(0.5, "points"))
```

## PCA

### All immunomodulator genes

```{r, warning=FALSE}
# convert expression df to matrix
immunomod_expr_mat <- immunomod_expr_df %>%
    dplyr::select(ParticipantBarcode, dplyr::contains("RNASeq")) %>% 
    tidyr::gather(gene_name, x, -ParticipantBarcode) %>% 
    dplyr::mutate(log_x = log(x + 1),
                  gene_name = stringr::str_replace(gene_name, "_RNASeq", "")) %>% 
    dplyr::select(gene_name, ParticipantBarcode, log_x) %>% 
    tidyr::spread(ParticipantBarcode, log_x) %>% 
    tibble::remove_rownames() %>% 
    tibble::column_to_rownames("gene_name") %>% 
    as.matrix()

immunomod_expr_pca <- immunomod_expr_mat %>% 
    t() %>% 
    prcomp()
```


#### Extract and format PC data

```{r, warning=FALSE}
pc_df <- immunomod_expr_pca %>% 
    broom::tidy("pcs")

immunomod_expr_pca_df <- immunomod_expr_pca %>% 
    broom::tidy("samples") %>% 
    dplyr::filter(PC <= 2) %>% 
    dplyr::left_join(irwg_fmx_df, by = c("row" = "ParticipantBarcode"))
```


#### PCA bi-plots

```{r}
pc1_label <- pc_df %>% 
    dplyr::filter(PC == 1) %>% 
    dplyr::transmute(label = sprintf("PC%s [%0.2f%%]", PC, 100*percent)) %>% 
    purrr::flatten_chr()

pc2_label <- pc_df %>% 
    dplyr::filter(PC == 2) %>% 
    dplyr::transmute(label = sprintf("PC%s [%0.2f%%]", PC, 100*percent)) %>% 
    purrr::flatten_chr()

p1 <- immunomod_expr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    ggplot2::ggplot(aes(x = `1`, y = `2`)) +
    ggplot2::geom_point(ggplot2::aes(fill = Subtype_Immune_Model_Based), 
                        size = 1.5, alpha = 0.8, shape = 21, colour = "black", 
                        stroke = 0.2) +
    ggplot2::xlab(pc1_label) +
    ggplot2::ylab(pc2_label) +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::theme(legend.position = "top") +
    ggplot2::coord_fixed()
p2 <- immunomod_expr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    ggplot2::ggplot(aes(x = `1`, y = `2`)) +
    ggplot2::geom_point(ggplot2::aes(fill = leukocyte_fraction), 
                        size = 1.5, alpha = 0.8, shape = 21, colour = "black", 
                        stroke = 0.2) +
    ggplot2::xlab(pc1_label) +
    ggplot2::ylab(pc2_label) +
    ggplot2::theme(legend.position = "top") +
    ggplot2::coord_fixed()
p1 + p2
```

```{r}
immunomod_expr_pca_df %>% 
    filter(PC == 1) %>% 
    ggplot(ggplot2::aes(x = leukocyte_fraction, y = value)) +
    ggplot2::geom_point(
        ggplot2::aes(fill = Subtype_Immune_Model_Based), 
        size = 1.5, alpha = 0.5,
        shape = 21, colour = "slategray", stroke = 0.2
    ) +
    ggplot2::geom_smooth(method = "lm") +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::guides(fill = FALSE) +
    ggplot2::ylab("PC1") +
    ggplot2::xlab("Leukocyte Fraction") +
    ggplot2::ggtitle("PC1 correlation with leukocyte fraction") +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8))
```


#### Bi-plots with loadings

```{r}
set.seed(0)
immunomod_factor_df <- immunomod_expr_pca_df %>% 
    spread(PC, value) %>% 
    select(row, leukocyte_fraction, Subtype_Immune_Model_Based) %>% 
    remove_rownames() %>% 
    column_to_rownames("row")

immunomod_factor_scores <- vegan::envfit(immunomod_expr_pca ~ 0 + ., 
                                 data = immunomod_factor_df, perm = 1000)

immunomod_arrow_df <- immunomod_factor_scores$factors$centroids %>% 
    as.data.frame() %>% 
    rownames_to_column("arrow") %>% 
    mutate(factor = str_extract(arrow, 
                                str_c(names(immunomod_factor_df), 
                                      collapse = "|")),
           arrow = str_replace(arrow, 
                               str_c(names(immunomod_factor_df), 
                                     collapse = "|"), 
                               "")) %>% 
    left_join(immunomod_factor_scores$factors$r %>% 
                  as.data.frame() %>% 
                  rownames_to_column() %>%
                  set_names(c("factor", "r2")),
              by = "factor") %>% 
    left_join(immunomod_factor_scores$factors$pvals %>% 
                  as.data.frame() %>% 
                  rownames_to_column() %>%
                  set_names(c("factor", "pval")),
              by = "factor") %>% 
    bind_rows(immunomod_factor_scores$vectors$arrows %>% 
    as.data.frame() %>% 
    rownames_to_column("arrow") %>% 
    mutate(factor = str_extract(arrow, 
                                str_c(names(immunomod_factor_df), 
                                      collapse = "|"))) %>% 
    left_join(immunomod_factor_scores$vectors$r %>% 
                  as.data.frame() %>% 
                  rownames_to_column() %>%
                  set_names(c("factor", "r2")),
              by = "factor") %>% 
    mutate(pval = immunomod_factor_scores$vectors$pvals)) %>% 
    mutate(xlength = PC1 * sqrt(r2),
           ylength = PC2 * sqrt(r2),
           label = ifelse(pval < 0.05, arrow, ""))
    
```

```{r}
multiplier <- 1
immunomod_expr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    ggplot2::ggplot(ggplot2::aes(x = `1`, y = `2`)) +
    ggplot2::stat_ellipse(geom = "polygon", alpha = 0.1) +
    ggplot2::coord_fixed() +
    ggplot2::geom_point(colour = "gray", size = 1.5, alpha = 0.5) +
    ggplot2::geom_segment(data = immunomod_arrow_df,
                          ggplot2::aes(x = 0, xend = multiplier*xlength, 
                                       y = 0, yend = multiplier*ylength),
                          arrow = arrow(length = unit(.2, "cm"), 
                                        type = "closed"),
                          colour = "black") +
    ggrepel::geom_text_repel(data = immunomod_arrow_df,
                             ggplot2::aes(x = multiplier*xlength, 
                                          y = multiplier*ylength, 
                                          label = label),
                             colour = "black", fontface = "bold", size = 4) +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::theme(legend.position = "top")

```

#### Per-subtype bi-plots

```{r, fig.height=7, fig.width=5}
bg_df <- immunomod_expr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    dplyr::select(-Subtype_Immune_Model_Based)

immunomod_expr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    ggplot2::ggplot(aes(x = `1`, y = `2`)) +
    ggplot2::geom_point(data = bg_df, 
                        colour = "gray", size = 1.5, alpha = 0.5) +
    ggplot2::geom_point(ggplot2::aes(fill = leukocyte_fraction), 
                        size = 1.5, alpha = 0.8, shape = 21, colour = "black", 
                        stroke = 0.2) +
    ggplot2::scale_fill_viridis_c() +
    ggplot2::guides(fill = guide_colourbar(title = "Leukocyte Fraction")) +
    ggplot2::xlab(pc1_label) +
    ggplot2::ylab(pc2_label) +
    ggplot2::coord_fixed() +
    ggplot2::facet_wrap(~ Subtype_Immune_Model_Based, ncol = 2) +
    ggplot2::ggtitle("PCA of immunomodulator expression across all tumor types") +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8),
                   legend.position = "bottom",
                   legend.direction = "horizontal")

# ggplot2::ggsave("../figures/Immunomod-SubtypePCA_20170720.pdf", p,
#        width = 6, height = 10, units = "cm", dpi = 300, scale = 2)
```

### Front-line immunomodulator genes

```{r}
# convert expression df to matrix
immunomod_sub_expr_mat <- immunomod_expr_df %>%
    dplyr::select(ParticipantBarcode, dplyr::contains("RNASeq")) %>% 
    
    tidyr::gather(gene_name, x, -ParticipantBarcode) %>%
    dplyr::mutate(gene_name = str_replace(gene_name, "_RNASeq", "")) %>% 
    dplyr::filter(gene_name %in% frontline_immunomod_genes) %>% 
    dplyr::mutate(log_x = log(x + 1)) %>%
    dplyr::select(gene_name, ParticipantBarcode, log_x) %>%
    tidyr::spread(ParticipantBarcode, log_x) %>%
    tibble::remove_rownames() %>%
    tibble::column_to_rownames("gene_name") %>%
    as.matrix()

immunomod_sub_expr_pca <- immunomod_sub_expr_mat %>%
    t() %>%
    prcomp()
```

#### Extract and format PC data

```{r, warning=FALSE}
pc_df <- immunomod_sub_expr_pca %>% 
    broom::tidy("pcs")

immunomod_sub_expr_pca_df <- immunomod_sub_expr_pca %>% 
    broom::tidy("samples") %>% 
    dplyr::filter(PC <= 2) %>% 
    dplyr::left_join(irwg_fmx_df, by = c("row" = "ParticipantBarcode"))
```


#### Plot top PCs

```{r, fig.height=7, fig.width=5, include=FALSE}
pc1_label <- pc_df %>% 
    dplyr::filter(PC == 1) %>% 
    dplyr::transmute(label = sprintf("PC%s [%0.2f%%]", PC, 100*percent)) %>% 
    purrr::flatten_chr()

pc2_label <- pc_df %>% 
    dplyr::filter(PC == 2) %>% 
    dplyr::transmute(label = sprintf("PC%s [%0.2f%%]", PC, 100*percent)) %>% 
    purrr::flatten_chr()

bg_df <- immunomod_sub_expr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    dplyr::select(-Subtype_Immune_Model_Based)

immunomod_sub_expr_pca_df %>% 
    tidyr::spread(PC, value) %>% 
    ggplot2::ggplot(aes(x = `1`, y = `2`)) +
    ggplot2::geom_point(data = bg_df, 
                        colour = "gray", size = 1.5, alpha = 0.5) +
    ggplot2::geom_point(ggplot2::aes(fill = Subtype_Immune_Model_Based), 
                        size = 1.5, alpha = 0.8, shape = 21, colour = "black", 
                        stroke = 0.2) +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::guides(fill = FALSE) +
    ggplot2::xlab(pc1_label) +
    ggplot2::ylab(pc2_label) +
    ggplot2::facet_wrap(~ Subtype_Immune_Model_Based, ncol = 2) +
    ggplot2::ggtitle("PCA of immunomodulator expression across all tumor types") +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8)) +
    ggplot2::coord_fixed()

# ggplot2::ggsave("../figures/Immunomod-SubtypePCA_20170720.pdf", p,
#        width = 6, height = 10, units = "cm", dpi = 300, scale = 2)
```

## Clustering

```{r, fig.height=10, fig.width=8}
df1 <- tibble::tibble(label = colnames(immunomod_expr_mat)) %>%
    dplyr::left_join(immunomod_expr_df, 
                     by = c("label" = "ParticipantBarcode")) %>% 
    select(Subtype_Immune_Model_Based, Study, leukocyte_fraction) %>% 
    as.data.frame()
ha <- HeatmapAnnotation(
    df1, 
    col = list(
        Subtype_Immune_Model_Based = c(
            "C1" = subtype_colors[1],
            "C2" = subtype_colors[2],
            "C3" = subtype_colors[3],
            "C4" = subtype_colors[4],
            "C5" = subtype_colors[5],
            "C6" = subtype_colors[6]
        ),
        Study = tcga_colors %>% 
            split(.$Disease) %>%
            purrr::map("Color") %>% 
            unlist(),
        leukocyte_fraction = colorRamp2(seq(0, 1, length.out = 11), 
                                        viridisLite::viridis(11))
    ), annotation_legend_param = list(
        Subtype_Immune_Model_Based = list(title = "Subtype"),
        Study = list(nrow = 6),
        leukocyte_fraction = list(title = "Leukocyte Fraction")
    )
)
df2 <- tibble::tibble(label = rownames(immunomod_expr_mat)) %>% 
    dplyr::left_join(immunomod_df, by = c("label" = "HGNC Symbol")) %>% 
    dplyr::mutate(`Immune Checkpoint` = dplyr::case_when(
        stringr::str_detect(`Immune Checkpoint`, "^I") ~ "Inhibitory",
        stringr::str_detect(`Immune Checkpoint`, "^S") ~ "Stimulatory",
        is.na(`Immune Checkpoint`) ~ "N/A",
        TRUE ~ `Immune Checkpoint`
    )) %>% 
    dplyr::select(`Immune Checkpoint`) %>% 
    as.data.frame()
ra <- rowAnnotation(
    df2,
    col = list(
        `Immune Checkpoint` = c(
            "Inhibitory" = "#333333",
            "Stimulatory" = "#E69F00",
            "N/A" = "#888888"
        )
    )
)
ht <- ComplexHeatmap::Heatmap(immunomod_expr_mat, name = "log(Expression)",
                        col = rev(RColorBrewer::brewer.pal(11, "RdBu")),
                        show_column_names = FALSE,
                        column_dend_height = unit(2, "cm"),
                        row_names_gp = gpar(fontsize = 8),
                        top_annotation = ha)
ComplexHeatmap::draw(ht + ra, annotation_legend_side = "bottom")
```



## Differential expression testing {#dge}

### One-way ANOVA for gene expression ~ subtype

```{r, include=TRUE}
immunomod_anova_df <- immunomod_zscore_df %>% 
    dplyr::group_by(gene_name) %>% 
    dplyr::do(
        broom::tidy(car::Anova(lm(log_x ~ 0 + Subtype_Immune_Model_Based, .),
                               white.adjust = TRUE))
    ) %>%
    dplyr::ungroup() %>% 
    dplyr::filter(term == "Subtype_Immune_Model_Based") %>%
    dplyr::arrange(dplyr::desc(statistic))

immunomod_anova_df[["p.adjust"]] <- p.adjust(immunomod_anova_df$p.value, 
                                             method = "BH")
immunomod_anova_df %>%
    dplyr::select(gene_name, term, statistic, p.value, p.adjust)
```

```{r}
immunomod_zscore_df %>% 
    filter(gene_name == "HMGB1") %>% 
    lm(log_x ~ 0 + Subtype_Immune_Model_Based, 
       data = .) -> model
hist(residuals(model))
plot(model)
```

```{r}
immunomod_zscore_df %>% 
    dplyr::left_join(irwg_fmx_df %>% 
                         select(ParticipantBarcode, leukocyte_fraction), 
                     by = "ParticipantBarcode") %>% 
    filter(gene_name == "HMGB1") %>% 
    lm(log_x ~ 0 + Subtype_Immune_Model_Based + leukocyte_fraction, 
       data = .) -> model
hist(residuals(model))
plot(model)
```

### Two-way ANOVA for gene expression ~ subtype + leukocyte fraction

```{r, include=TRUE}
immunomod_lf_anova_df <- immunomod_zscore_df %>% 
    dplyr::left_join(irwg_fmx_df %>% 
                         select(ParticipantBarcode, leukocyte_fraction), 
                     by = "ParticipantBarcode") %>% 
    dplyr::group_by(gene_name) %>% 
    dplyr::do(
        broom::tidy(car::Anova(lm(log_x ~ 0 + Subtype_Immune_Model_Based + leukocyte_fraction, .),
                               white.adjust = TRUE))
    ) %>%
    dplyr::ungroup() %>% 
    dplyr::filter(term == "Subtype_Immune_Model_Based") %>%
    dplyr::arrange(dplyr::desc(statistic))

immunomod_lf_anova_df[["p.adjust"]] <- p.adjust(immunomod_lf_anova_df$p.value, 
                                             method = "BH")
immunomod_lf_anova_df %>%
    dplyr::select(gene_name, term, statistic, p.value, p.adjust)
```

### Top genes from two-way ANOVA

Picking a few of the top genes (based on test statistic... as p-value isn't very useful), I can see how their distributions are changing across clusters. There certainly seems to be a reasonable difference, but I don't know how interesting this is.

```{r}
top_genes <- immunomod_lf_anova_df %>% 
    dplyr::slice(1:3) %>% 
    .[["gene_name"]]
p <- immunomod_zscore_df %>% 
    dplyr::filter(gene_name %in% top_genes) %>%
    dplyr::mutate(gene_name = factor(gene_name, levels = top_genes)) %>% 
    ggplot2::ggplot(aes(y = log_x, x = Subtype_Immune_Model_Based)) +
    ggbeeswarm::geom_quasirandom(
        ggplot2::aes(fill = Subtype_Immune_Model_Based), 
        size = 1, alpha = 0.5,
        shape = 21, colour = "slategray", stroke = 0.2
    ) +
    ggplot2::geom_boxplot(
        ggplot2::aes(colour = Subtype_Immune_Model_Based), 
        alpha = 0.6, outlier.size = 0, outlier.alpha = 0
    ) +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::scale_colour_manual(values = subtype_colors) +
    ggplot2::guides(colour = FALSE, fill = FALSE) +
    ggplot2::ylab("log(Expression)") +
    ggplot2::xlab("Subtype") +
    ggplot2::facet_wrap(~ gene_name, ncol = 3) +
    ggplot2::ggtitle("Immunomodulator expression across subtypes") +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8))
p
# ggsave("../figures/Immunomod-SubtypeDGE_expressionDists_20170720.pdf", p,
#        width = 10, height = 4.5, units = "cm", dpi = 300, scale = 2)
```

```{r}
p <- immunomod_zscore_df %>% 
    dplyr::left_join(irwg_fmx_df %>% 
                         select(ParticipantBarcode, leukocyte_fraction), 
                     by = "ParticipantBarcode") %>% 
    dplyr::filter(gene_name %in% top_genes) %>%
    dplyr::mutate(gene_name = factor(gene_name, levels = top_genes)) %>% 
    ggplot2::ggplot(ggplot2::aes(y = log_x, x = leukocyte_fraction)) +
    ggplot2::geom_point(
        ggplot2::aes(fill = Subtype_Immune_Model_Based), 
        size = 1.5, alpha = 0.5,
        shape = 21, colour = "slategray", stroke = 0.2
    ) +
    ggplot2::geom_smooth(method = "lm") +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::guides(fill = FALSE) +
    ggplot2::ylab("log(Expression)") +
    ggplot2::xlab("Leukocyte Fraction") +
    ggplot2::facet_wrap(~ gene_name) +
    ggplot2::ggtitle("Immunomodulator correlation with leukocyte fraction") +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8))
p

# ggsave("../figures/Immunomod-SubtypeDGE_CorrLF_20170720.pdf", p,
#        width = 10, height = 4.5, units = "cm", dpi = 300, scale = 2)
```

```{r}
immunomod_lfresid_df <- immunomod_zscore_df %>% 
    dplyr::left_join(irwg_fmx_df %>% 
                         select(ParticipantBarcode, leukocyte_fraction), 
                     by = "ParticipantBarcode") %>% 
    tidyr::nest(-gene_name) %>% 
    dplyr::mutate(
        fit = purrr::map(data, ~ lm(log_x ~ leukocyte_fraction, data = .)),
        results = purrr::map(fit, broom::augment)
    ) %>% 
    tidyr::unnest(data, results)
```

```{r}
p <- immunomod_lfresid_df %>% 
    dplyr::filter(gene_name %in% top_genes) %>%
    dplyr::mutate(gene_name = factor(gene_name, levels = top_genes)) %>% 
    ggplot2::ggplot(aes(y = .resid, x = Subtype_Immune_Model_Based)) +
    ggbeeswarm::geom_quasirandom(
        ggplot2::aes(fill = Subtype_Immune_Model_Based), 
        size = 1, alpha = 0.5,
        shape = 21, colour = "slategray", stroke = 0.2
    ) +
    ggplot2::geom_boxplot(
        ggplot2::aes(colour = Subtype_Immune_Model_Based), 
        alpha = 0.6, outlier.size = 0, outlier.alpha = 0
    ) +
    ggplot2::scale_fill_manual(values = subtype_colors) +
    ggplot2::scale_colour_manual(values = subtype_colors) +
    ggplot2::guides(colour = FALSE, fill = FALSE) +
    ggplot2::ylab("LF Residual Expression") +
    ggplot2::xlab("Subtype") +
    ggplot2::facet_wrap(~ gene_name, ncol = 3) +
    ggplot2::ggtitle("Immunomodulator expression across subtypes") +
    ggplot2::theme(title = ggplot2::element_text(size = 8),
                   axis.title.x = ggplot2::element_text(size = 8),
                   axis.title.y = ggplot2::element_text(size = 8))
p
```

