---
title: "PanCan IRWG Tumor Clonality Correlative Analysis"
author: "James A. Eddy"
date: '`r lubridate::today()`'
output:
  html_notebook:
    code_folding: hide
    fig_width: 8
    toc: yes
    toc_float: yes
---

# Summary

Exploring associations of tumor clonality across tumor types in TCGA with suggested correlates defined by the Immune Response Working Group:

+ Overall Leukocyte fraction  
+ Individual Relative CIBERSORT fraction  
+ Mutation Load  
+ TCR,BCR Diversity  
+ Gene expression


```{r setup, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE, cache=FALSE}
# Synapse client
library(synapseClient)

# parallel computing
library(parallel)

# viz packages
library(ggthemes)
library(viridis)
library(ggbeeswarm)

# packages for general data munging, formatting
library(feather)
library(stringr)
library(forcats)
library(broom)
library(tidyverse)

my_theme_bw <- function() {
    theme_bw() +
        theme(axis.title = element_text(face = "bold", size = rel(0.9)),
              legend.title = element_text(face = "bold"),
              plot.title = element_text(face = "bold"))
}
ggplot2::theme_set(my_theme_bw())
scale_fill_discrete <- function(...) scale_fill_colorblind(...)
scale_fill_continuous <- function(...) scale_fill_viridis(...)
scale_colour_discrete <- function(...) scale_color_colorblind(...)
scale_colour_continuous <- function(...) scale_color_viridis(...)
```

-----

\  

# Prepare data

## Retrieve clonality data from Synapse

Data are stored on Synapse in the folder `syn4602485` for "Aneuploidy and Copy number".

Aneuploidy and Copy number: syn4602485
syn7300639 ("Purity_Ploidy_All_Samples_9_28_16.tsv")
filename: "TCGA_mastercalls.abs_tables_JSedit.fixed.txt "

```{r}
if (!dir.exists("../data")) {
    dir.create("../data")
}
synapseLogin()

# file_data <- synGet("syn7300639", downloadLocation = "./data/tcga")
clonality_file <- "../data/tcga/TCGA_mastercalls.abs_tables_JSedit.fixed.txt"
clonality_df <- read_tsv(clonality_file)
```

## Sample filtering

Sample Quality Annotations:
syn4551248 ("merged_sample_quality_annotations.tsv")

```{r}
sample_qual_file <- synGet("syn4551248", downloadLocation = "../data/tcga/")
sample_qual_df <- sample_qual_file %>% 
    getFileLocation() %>% 
    read_tsv()
```

Remove samples based on `Do_not_use=True`, and remove cases with `AWG_excluded_because_of_pathology=True`. 

```{r}
# samples to exclude from all datasets
exclude_samples <- sample_qual_df %>% 
    mutate(AWG_excluded_because_of_pathology = parse_logical(AWG_excluded_because_of_pathology),
           Do_not_use = parse_logical(str_to_upper(Do_not_use))) %>% 
    filter(AWG_excluded_because_of_pathology | Do_not_use)
```

Remove samples from clonality dataset, add disease types. Also remove samples with missing values for subclonal faction. Add additional column to store vial ID (to ease mapping between data sets).

```{r}
clonality_corr_df <- clonality_df %>% 
    filter(!(sample %in% exclude_samples$aliquot_barcode),
           !is.na(`Subclonal genome fraction`)) %>% 
    left_join(sample_qual_df %>% 
                  select(aliquot_barcode, `cancer type`),
              by = c("sample" = "aliquot_barcode")) %>% 
    mutate(vial_id = str_replace(sample, "(\\-[:alnum:]+){3}$", ""))
```

\  

## Set up plotting colors

```{r}
tcga_colors <- tribble(
    ~Color, ~Disease,
    "#ED2891", "BRCA",
    "#B2509E", "GBM",
    "#D49DC7", "LGG",
    "#C1A72F", "ACC",
    "#E8C51D", "PCPG",
    "#F9ED32", "THCA",
    "#CACCDB", "CHOL",
    "#9EDDF9", "COAD",
    "#007EB5", "ESCA",
    "#104A7F", "LIHC",
    "#6E7BA2", "PAAD",
    "#DAF1FC", "READ",
    "#00AEEF", "STAD",
    "#F6B667", "CESC",
    "#D97D25", "OV",
    "#F89420", "UCEC",
    "#FBE3C7", "UCS",
    "#009444", "HNSC",
    "#97D1A9", "UVM",
    "#754C29", "LAML",
    "#CEAC8F", "THYM",
    "#3953A4", "DLBC",
    "#BBD642", "SKCM",
    "#00A99D", "SARC",
    "#542C88", "LUAD",
    "#A084BD", "LUSC",
    "#D3C3E0", "MESO",
    "#FAD2D9", "BLCA",
    "#EA7075", "KICH",
    "#ED1C24", "KIRC",
    "#F8AFB3", "KIRP",
    "#7E1918", "PRAD",
    "#BE1E2D", "TGCT"
)

clonality_corr_df <- clonality_corr_df %>% 
    dplyr::rename(Disease = `cancer type`) %>% 
    mutate(Disease = factor(Disease, levels = tcga_colors$Disease))
```

-----

\  

# Explore clonality data

## Sample characteristics

Note: between the clonality and sample quality data, I don't have any clear indication of sample type. I may want to use another source (e.g., the miRNA sample data) to filter for primary tumors only.

\  

## Subclonal fraction values

I can look at the distribution of subclonal fraction values with boxplots:

```{r}
clonality_corr_df %>% 
    ggplot(aes(x = Disease, y = `Subclonal genome fraction`)) +
    geom_boxplot(aes(fill = Disease), alpha = 1, outlier.size = 0.5) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_fill_manual(values = tcga_colors$Color) +
    guides(fill = FALSE)
```

-----

\  

# Prepare correlate data

+ Overall Leukocyte fraction  
+ Individual Relative CIBERSORT fraction  
+ Mutation Load  
+ TCR,BCR Diversity  
+ Gene Expression

## Leukocyte fraction

### Retrieve/load data

Cellular Content: syn7994728
syn5808205 ("TCGA_all_leuk_estimate.masked.20170107.tsv")

```{r}
# file_data <- synGet("syn5808205", downloadLocation = "./")
leuk_frac_file <- "../data/tcga/TCGA_all_leuk_estimate.masked.20170107.tsv"
leuk_frac_df <- read_tsv(leuk_frac_file, col_names = FALSE) %>% 
    set_names(c("disease", "id", "leuk_frac"))
```

### Sample filtering

```{r}
leuk_frac_df <- leuk_frac_df %>%
    filter(!(id %in% exclude_samples$aliquot_barcode))
```

### Sample matching

Identify matched samples between miRNA and leukocyte fraction.

```{r}
leuk_frac_ids <- leuk_frac_df %>% 
    select(id) %>%
    mutate(vial_id = str_replace(id, "(\\-[:alnum:]+){3}$", "")) %>% 
    arrange()

clonality_ids <- clonality_corr_df %>%
    select(id = sample, vial_id) %>%
    arrange()

clonality_leuk_frac_shared_ids <- inner_join(clonality_ids, leuk_frac_ids, 
                                         by = "vial_id",
                                         suffix = c("_clonality", "_leuk_frac"))

# only keep samples with matched vial ID AND portion number
portion_id_minus_analyte_regex <- "([:alnum:]+\\-){4}[0-9]+"
# plate_id_only_regex <- "[:alnum:]+(?=([:alnum:]\\-[:alnum:]+){1}$)"
clonality_leuk_frac_shared_ids <- clonality_leuk_frac_shared_ids %>%
    filter((str_extract(id_clonality, portion_id_minus_analyte_regex)
            == str_extract(id_leuk_frac, portion_id_minus_analyte_regex)))
```

NOTE: several samples were assayed on multiple plates; average the leukocyte fraction across these before computing correlations with miRNA

### Correlate data formatting

```{r}
leuk_frac_corr_file <- "../data/leuk_frac_correlates_for_clonality.feather"
force_format <- TRUE
if (!file.exists(leuk_frac_corr_file) | force_format) {
    leuk_frac_corr_df <- clonality_leuk_frac_shared_ids %>% 
        left_join(leuk_frac_df, by = c("id_leuk_frac" = "id")) %>% 
        group_by(disease, vial_id) %>% 
        summarise(leuk_frac = mean(leuk_frac)) %>% 
        ungroup()
    
    write_feather(leuk_frac_corr_df, leuk_frac_corr_file)
} else {
    leuk_frac_corr_df <- read_feather(leuk_frac_corr_file)
}
```

### Disease-wise correlations

```{r warning=FALSE}
if (!dir.exists("../results")) {
    dir.create("../results")
}
clonality_leuk_frac_corr_file <- "../results/clonality_leuk_frac_correlation.feather"
force_compute <- FALSE
if (!file.exists(clonality_leuk_frac_corr_file) | force_compute) {
    # skip immune cell cancers?
    d_list <- clonality_corr_df %>% 
        # filter((!Disease %in% c("LAML", "THYM", "DLBC"))) %>% 
        filter(Disease %in% leuk_frac_corr_df$disease) %>% 
        distinct(Disease) %>% 
        mutate(Disease = as.character(Disease)) %>% 
        .$Disease %>%
        set_names(.) %>% 
        as.list()
    
    d <- d_list[1]
    corr_df_list <- mclapply(d_list, mc.cores = 4, function(d) {
        samples_d <- clonality_corr_df %>%
            filter(Disease == d) %>%
            select(vial_id) %>%
            flatten_chr() %>%
            intersect(leuk_frac_corr_df$vial_id)

        source_df <- clonality_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            select(sample = vial_id, x = `Subclonal genome fraction`) %>%
            filter(!is.na(x)) %>% 
            mutate(clonality = "subclonal_frac")
        
        target_df <- leuk_frac_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            dplyr::rename(sample = vial_id, x = leuk_frac) %>% 
            mutate(correlate = "leuk_frac")
        
        corr_df <- inner_join(source_df, target_df,
                              by = "sample",
                              suffix = c("_source", "_target")) %>%
            group_by(clonality, correlate) %>%
            do(tidy(cor.test(.$x_source, .$x_target, method = "spearman"))) %>% 
            ungroup()
        corr_df[["p.adjust"]] <- p.adjust(corr_df$p.value, method = "BH")
        return(corr_df)
    })
    
    clonality_leuk_frac_corr_df <- bind_rows(corr_df_list, .id = "disease") %>% 
        mutate(correlate_type = "leukocyte fraction")
    write_feather(clonality_leuk_frac_corr_df, clonality_leuk_frac_corr_file)
} else {
    clonality_leuk_frac_corr_df <- read_feather(clonality_leuk_frac_corr_file)
}
```

\  

## CIBERSORT fraction

### Retrieve/load data

Cellular Content: syn4991611
syn8024565 ("TCGA.cluster-by-CIBERSORT-relative.tsv")

or...?

syn7337221 ("TCGA.Kallisto.fullIDs.cibersort.relative.tsv")

```{r}
# ciber_frac_file <- "../data/tcga/TCGA.cluster-by-CIBERSORT-relative.tsv"
ciber_frac_file <- "../data/tcga/TCGA.Kallisto.fullIDs.cibersort.relative.tsv"

ciber_frac_df <- read_tsv(ciber_frac_file)
```

### Sample filtering

```{r}
ciber_frac_df <- ciber_frac_df %>%
    mutate(id = str_replace_all(SampleID, "\\.", "\\-")) %>% 
    filter(!(id %in% exclude_samples$aliquot_barcode))
```

### Sample matching

Identify matched samples between miRNA and CIBERSORT fraction.

```{r}
ciber_frac_ids <- ciber_frac_df %>% 
    select(id) %>%
    mutate(vial_id = str_replace(id, "(\\-[:alnum:]+){3}$", "")) %>% 
    arrange()

# only keep samples with matched vial ID AND portion number
clonality_ciber_frac_shared_ids <- inner_join(
    clonality_ids, ciber_frac_ids,
    by = "vial_id", suffix = c("_clonality", "_ciber_frac")
    ) %>% 
    distinct() %>% 
    filter((str_extract(id_clonality, portion_id_minus_analyte_regex) 
            == str_extract(id_ciber_frac, portion_id_minus_analyte_regex)))
```

NOTE: several samples were assayed on multiple plates; average the CIBERSORT fraction across these before computing correlations with miRNA

### Correlate data formatting

```{r}
ciber_frac_corr_file <- "../data/ciber_frac_correlates_for_clonality.feather"
force_format <- FALSE
if (!file.exists(ciber_frac_corr_file) | force_format) {
    ciber_frac_corr_df <- clonality_ciber_frac_shared_ids %>%
        left_join(ciber_frac_df, by = c("id_ciber_frac" = "id")) %>% 
        select(-id_clonality, -id_ciber_frac, 
               -SampleID, -P.value, -Correlation, -RMSE) %>% 
        group_by(CancerType, vial_id) %>%
        summarise_each(funs(mean)) %>% 
        ungroup()
    
    write_feather(ciber_frac_corr_df, ciber_frac_corr_file)
} else {
    ciber_frac_corr_df <- read_feather(ciber_frac_corr_file)
}
```

### Disease-wise correlations

```{r warning=FALSE}
clonality_ciber_frac_corr_file <- "../results/clonality_ciber_frac_correlation.feather"
force_compute <- FALSE
if (!file.exists(clonality_ciber_frac_corr_file) | force_compute) {
    # skip immune cell cancers?
    d_list <- clonality_corr_df %>% 
        # filter((!Disease %in% c("LAML", "THYM", "DLBC"))) %>% 
        filter(Disease %in% ciber_frac_corr_df$CancerType) %>% 
        distinct(Disease) %>% 
        mutate(Disease = as.character(Disease)) %>% 
        .$Disease %>%
        set_names(.) %>% 
        as.list()
    
    corr_df_list <- mclapply(d_list, mc.cores = 4, function(d) {
        samples_d <- clonality_corr_df %>%
            filter(Disease == d) %>%
            select(vial_id) %>%
            flatten_chr() %>%
            intersect(ciber_frac_corr_df$vial_id)

        source_df <- clonality_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            select(sample = vial_id, x = `Subclonal genome fraction`) %>%
            filter(!is.na(x)) %>% 
            mutate(clonality = "subclonal_frac")

        target_df <- ciber_frac_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            select(-CancerType) %>% 
            gather(correlate, x, -vial_id) %>% 
            dplyr::rename(sample = vial_id)
        
        corr_df <- inner_join(source_df, target_df,
                              by = "sample",
                              suffix = c("_source", "_target")) %>%
            group_by(clonality, correlate) %>%
            do(tidy(cor.test(.$x_source, .$x_target, method = "spearman"))) %>% 
            ungroup()
        corr_df[["p.adjust"]] <- p.adjust(corr_df$p.value, method = "BH")
        return(corr_df)
    })
    
    clonality_ciber_frac_corr_df <- bind_rows(corr_df_list, .id = "disease") %>% 
        mutate(correlate_type = "CIBERSORT fraction")
    write_feather(clonality_ciber_frac_corr_df, clonality_ciber_frac_corr_file)
} else {
    clonality_ciber_frac_corr_df <- read_feather(clonality_ciber_frac_corr_file)
}
```

\  

## Mutation load

### Retrieve/load data

Mutation Load: syn5706827
syn7994728 ("mutation-load")

* exclude LAML?

```{r}
mut_load_file <- "../data/tcga/mutation-load-vial.txt"
mut_load_df <- read_tsv(mut_load_file)
```

### Sample filtering

Can't filter aliquots, as data only includes IDs specific to the level of vial

### Sample matching

Identify matched samples between miRNA and mutational load.

```{r}
mut_load_ids <- mut_load_df %>% 
    select(id = Tumor_Sample_ID_Vial) %>% 
    mutate(vial_id = id) %>%
    arrange()

clonality_mut_load_shared_ids <- inner_join(
    clonality_ids, mut_load_ids, 
    by = "vial_id", suffix = c("_clonality", "_mut_load")
    )
```

### Correlate data formatting

```{r}
mut_load_corr_file <- "../data/mut_load_correlates_for_clonality.feather"
force_format <- FALSE
if (!file.exists(mut_load_corr_file) | force_format) {
    mut_load_corr_df <- clonality_mut_load_shared_ids %>% 
        left_join(mut_load_df, by = c("id_mut_load" = "Tumor_Sample_ID_Vial"))
    
    write_feather(mut_load_corr_df, mut_load_corr_file)
} else {
    mut_load_corr_df <- read_feather(mut_load_corr_file)
}
```

### Disease-wise correlations

```{r warning=FALSE}
clonality_mut_load_corr_file <- "../results/clonality_mut_load_correlation.feather"
force_compute <- FALSE
if (!file.exists(clonality_mut_load_corr_file) | force_compute) {
    # skip immune cell cancers?
    d_list <- clonality_corr_df %>% 
        # filter((!Disease %in% c("LAML", "THYM", "DLBC"))) %>% 
        filter(Disease %in% mut_load_corr_df$cohort) %>% 
        distinct(Disease) %>% 
        mutate(Disease = as.character(Disease)) %>% 
        .$Disease %>%
        set_names(.) %>% 
        as.list()
    
    corr_df_list <- mclapply(d_list, mc.cores = 4, function(d) {
        samples_d <- clonality_corr_df %>%
            filter(Disease == d) %>%
            select(vial_id) %>%
            flatten_chr() %>%
            intersect(mut_load_corr_df$vial_id)

        source_df <- clonality_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            select(sample = vial_id, x = `Subclonal genome fraction`) %>%
            filter(!is.na(x)) %>% 
            mutate(clonality = "subclonal_frac")

        target_df <- mut_load_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            select(-id_clonality, -id_mut_load, -cohort, -Patient_ID, 
                   -Tumor_Sample_ID) %>% 
            gather(correlate, x, -vial_id) %>% 
            dplyr::rename(sample = vial_id)
        
        corr_df <- inner_join(source_df, target_df,
                              by = "sample",
                              suffix = c("_source", "_target")) %>%
            group_by(clonality, correlate) %>%
            do(tidy(cor.test(.$x_source, .$x_target, method = "spearman"))) %>% 
            ungroup()
        corr_df[["p.adjust"]] <- p.adjust(corr_df$p.value, method = "BH")
        return(corr_df)
    })
    
    clonality_mut_load_corr_df <- bind_rows(corr_df_list, .id = "disease") %>% 
        mutate(correlate_type = "mutation load")
    write_feather(clonality_mut_load_corr_df, clonality_mut_load_corr_file)
} else {
    clonality_mut_load_corr_df <- read_feather(clonality_mut_load_corr_file)
}
```

\  

## TCR/BCR diversity

### Retrieve/load data

T Cell Receptor / Brown et al: syn5876488
syn7063422 ("mitcr_sampleStatistics_20160714.tsv")

```{r}
tcr_div_file <- "../data/tcga/mitcr_sampleStatistics_20160714.tsv"

tcr_div_df <- read_tsv(tcr_div_file)
```

### Sample filtering

```{r}
tcr_dv_df <- tcr_div_df %>%
    filter(!(AliquotBarcode %in% exclude_samples$aliquot_barcode))
```

### Sample matching

Identify matched samples between miRNA and CIBERSORT fraction.

```{r}
tcr_div_ids <- tcr_div_df %>% 
    select(id = AliquotBarcode, vial_id = SampleBarcode) %>% 
    arrange()

# only keep samples with matched vial ID AND portion number
clonality_tcr_div_shared_ids <- inner_join(
    clonality_ids, tcr_div_ids,
    by = "vial_id", suffix = c("_clonality", "_tcr_div")
    ) %>% 
    distinct() %>% 
    filter((str_extract(id_clonality, portion_id_minus_analyte_regex) 
            == str_extract(id_tcr_div, portion_id_minus_analyte_regex)))
```

NOTE: several samples were assayed on multiple plates; average the TCR diversity across these before computing correlations with miRNA

### Correlate data formatting

Some samples (`AliquotBarcode`) have 2 values for Shannon entropy, apparently corresponding to separate analyses on CGHub. I'll go ahead and average these values per aliquot, prior to averaging Shannon values per vial ID. As a small measure of QC, I'll also only keep observations with at least 1 TCRA read or 1 TCRB read.

```{r}
tcr_div_corr_file <- "../data/tcr_div_correlates_for_clonality.feather"
force_format <- FALSE
if (!file.exists(tcr_div_corr_file) | force_format) {
    tcr_div_corr_df <- clonality_tcr_div_shared_ids %>%
        left_join(tcr_div_df, by = c("id_tcr_div" = "AliquotBarcode")) %>%
        select(-CGHub_analysis_id) %>% 
        distinct() %>% 
        filter(totTCRa_reads > 0 | totTCRb_reads > 0) %>%
        group_by(Study, vial_id, id_tcr_div) %>%
        summarize(shannon = mean(shannon)) %>%
        group_by(Study, vial_id) %>%
        summarise(shannon = mean(shannon)) %>% 
        ungroup()

    write_feather(tcr_div_corr_df, tcr_div_corr_file)
} else {
    tcr_div_corr_df <- read_feather(tcr_div_corr_file)
}
```

### Disease-wise correlations

```{r warning=FALSE}
clonality_tcr_div_corr_file <- "../results/clonality_tcr_div_correlation.feather"
force_compute <- FALSE
if (!file.exists(clonality_tcr_div_corr_file) | force_compute) {
    # skip immune cell cancers?
    d_list <- clonality_corr_df %>% 
        # filter((!Disease %in% c("LAML", "THYM", "DLBC"))) %>% 
        filter(Disease %in% tcr_div_corr_df$Study) %>% 
        distinct(Disease) %>% 
        mutate(Disease = as.character(Disease)) %>% 
        .$Disease %>%
        set_names(.) %>% 
        as.list()
    
    corr_df_list <- mclapply(d_list, mc.cores = 4, function(d) {
        samples_d <- clonality_corr_df %>%
            filter(Disease == d) %>%
            select(vial_id) %>%
            flatten_chr() %>%
            intersect(tcr_div_corr_df$vial_id)

        source_df <- clonality_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            select(sample = vial_id, x = `Subclonal genome fraction`) %>%
            filter(!is.na(x)) %>% 
            mutate(clonality = "subclonal_frac")

        target_df <- tcr_div_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            select(-Study) %>% 
            dplyr::rename(sample = vial_id) %>% 
            gather(correlate, x, -sample)
        
        corr_df <- inner_join(source_df, target_df,
                              by = "sample",
                              suffix = c("_source", "_target")) %>%
            group_by(clonality, correlate) %>%
            do(tidy(cor.test(.$x_source, .$x_target, method = "spearman"))) %>% 
            ungroup()
        corr_df[["p.adjust"]] <- p.adjust(corr_df$p.value, method = "BH")
        return(corr_df)
    })
    
    clonality_tcr_div_corr_df <- bind_rows(corr_df_list, .id = "disease") %>% 
        mutate(correlate_type = "TCR diversity")
    write_feather(clonality_tcr_div_corr_df, clonality_tcr_div_corr_file)
} else {
    clonality_tcr_div_corr_df <- read_feather(clonality_tcr_div_corr_file)
}
```

\  

## Gene expression

+ Expression of  
    + IFNG,IFN-gamma  
    + PRF1,Perforin  
    + GZMA,Granzyme A  
    + PDCD1,PD-1  
    + CD274,PD-L1  
    + PDCD1LG2,PD-L2  
    + IL10,IL-10  
    + TGFB1,TGF-beta  
    + IDO1,IDO  
    + HLA-A  

### Retrieve/load mRNA data

Batch effects normalized mRNA data: syn4976363
syn4976369 ("EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv")
syn4976366 (" EB++GeneExpAnnotation.tsv")

```{r include=FALSE}
mrna_synfolder <- "syn4976363"
mrna_synfiles <- synapseQuery(
    sprintf('select * from file where parentId=="%s"', mrna_synfolder)
)

# download files and store data/paths in new data frame
mrna_files <- mrna_synfiles %>% 
    mutate(file_data = map(file.id, function(synid) {
        synGet(synid, downloadLocation = "../data/tcga")
    }),
    file_path = map_chr(file_data, getFileLocation))
```

Sample characteristics are stored in a tab-delimited text file (Synapse ID: `syn4976366`) and can be loaded with `read_tsv()`.

```{r}
# load sample data
mrna_sample_file <- mrna_files %>% 
    filter(file.id == "syn4976366") %>% 
    .[["file_path"]]
mrna_sample_df <- read_tsv(mrna_sample_file)
```

### Sample filtering

Remove samples from mRNA dataset.

```{r}
mrna_sample_df <- mrna_sample_df %>% 
    filter(!(SampleID %in% exclude_samples$aliquot_barcode))
```

### Sample matching

Identify matched samples between miRNA and mRNA.

```{r}
mrna_ids <- mrna_sample_df %>% 
    select(SampleID) %>%
    mutate(vial_id = str_replace(SampleID, "(\\-[:alnum:]+){3}$", "")) %>% 
    arrange()

clonality_mrna_shared_ids <- inner_join(clonality_ids, mrna_ids, by = "vial_id")

# only keep samples with matched vial ID AND portion number
clonality_mrna_shared_ids <- clonality_mrna_shared_ids %>%
    filter(str_extract(id, portion_id_minus_analyte_regex)
           == str_extract(SampleID, portion_id_minus_analyte_regex))
```

mRNA normalized, batch corrected expression values for all samples are stored as a matrix in a TSV file (Synapse ID: `syn4976369`) and can be loaded with `read_tsv()`.

### Correlate data formatting

List of genes accessed [here](https://docs.google.com/spreadsheets/d/1aqOXYsU1ubkbxIZI_5p8ZRootgOAT0KweMA3LvSZ7HY/edit#gid=0) and saved as a TSV at `data/Cancer Immunomodulators - TCGA PanImmune Group - Direct Relationship.tsv`:

```{r}
gene_correlate_file <- "../data/Cancer Immunomodulators - TCGA PanImmune Group - Direct Relationship.tsv"
gene_correlate_df <- read_tsv(gene_correlate_file)
```


```{r}
mrna_corr_file <- "../data/mrna_correlates_for_clonality.feather"
force_format <- FALSE # loading full expression data is slow; avoid if possible
if (!file.exists(mrna_corr_file) | force_format) {
    # load normalized, batch-corrected expression data
    mrna_norm_file <- mrna_files %>% 
        filter(file.id == "syn4976369") %>% 
        .[["file_path"]]
    mrna_norm_df <- read_tsv(mrna_norm_file, progress = FALSE)
    
    # gene_list <- c("IFNG", "PRF1", "GZMA", "PDCD1", "CD274", "PDCD1LG2", "IL10", 
    #                "TGFB1", "IDO1", "HLA-A")  
    mrna_corr_df <- mrna_norm_df %>% 
        separate(gene_id, c("gene_name", "gene_id"), sep = "\\|") %>% 
        filter((gene_id %in% gene_correlate_df$`Entrez ID`) 
               | (gene_name %in% gene_correlate_df$`HGNC Symbol`)) %>% 
        select(one_of(c("gene_name", "gene_id", 
                        clonality_mrna_shared_ids$SampleID)))
    write_feather(mrna_corr_df, mrna_corr_file)
} else {
    mrna_corr_df <- read_feather(mrna_corr_file)
}
```

### Disease-wise correlations

```{r message=FALSE, warning=FALSE}
clonality_mrna_corr_file <- "../results/clonality_mrna_correlation.feather"
force_compute <- FALSE
if (!file.exists(clonality_mrna_corr_file) | force_compute) {
    # skip immune cell cancers?
    d_list <- clonality_corr_df %>% 
        # filter((!Disease %in% c("LAML", "THYM", "DLBC"))) %>% 
        distinct(Disease) %>% 
        mutate(Disease = as.character(Disease)) %>% 
        .$Disease %>%
        set_names(.) %>% 
        as.list()
    
    corr_df_list <- mclapply(d_list, mc.cores = 4, function(d) {
        samples_d <- clonality_corr_df %>%
            filter(Disease == d) %>%
            select(vial_id) %>%
            flatten_chr() %>%
            intersect(str_replace(names(mrna_corr_df),
                                  "(\\-[:alnum:]+){3}$", ""))
        source_df <- clonality_corr_df %>%
            filter(vial_id %in% samples_d) %>% 
            select(sample = vial_id, x = `Subclonal genome fraction`) %>%
            filter(!is.na(x)) %>% 
            mutate(clonality = "subclonal_frac")
        
        target_df <- mrna_corr_df %>%
            set_names(str_replace(names(.), "(\\-[:alnum:]+){3}$", "")) %>%
            select(one_of(c("gene_name", samples_d))) %>%
            dplyr::rename(correlate = gene_name) %>%
            gather(sample, x, -correlate) %>% 
            filter(!is.na(x))
        
        corr_df <- inner_join(source_df, target_df,
                              by = "sample",
                              suffix = c("_source", "_target")) %>%
            group_by(clonality, correlate) %>%
            do(tidy(cor.test(.$x_source, .$x_target, method = "spearman"))) %>% 
            ungroup()
        corr_df[["p.adjust"]] <- p.adjust(corr_df$p.value, method = "BH")
        return(corr_df)
    })
    
    clonality_mrna_corr_df <- bind_rows(corr_df_list, .id = "disease") %>% 
        mutate(correlate_type = "mRNA expression")
    write_feather(clonality_mrna_corr_df, clonality_mrna_corr_file)
} else {
    clonality_mrna_corr_df <- read_feather(clonality_mrna_corr_file)
}
```

-----

\  

# Summarize results

```{r, fig.width=5, fig.height=3}
bind_rows(clonality_leuk_frac_corr_df, clonality_ciber_frac_corr_df, 
          clonality_mut_load_corr_df, clonality_tcr_div_corr_df,
          clonality_mrna_corr_df) %>%
    group_by(disease, correlate_type) %>% 
    summarise(num_correlations = length(estimate),
              significant = sum(p.adjust < 0.05, na.rm = TRUE),
              other = num_correlations - significant) %>% 
    gather(correlations, total, -disease, -correlate_type, -num_correlations) %>% 
    ggplot(aes(x = disease, y = total)) +
    geom_col(aes(fill = disease, alpha = correlations), colour = "slategray") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          axis.text.y = element_text(angle = 45),
          legend.position = "top") +
    scale_fill_manual(values = tcga_colors$Color) +
    scale_alpha_manual(values = c(0.4, 1)) +
    guides(fill = FALSE) +
    facet_wrap(~ correlate_type, ncol = 2, scales = "free_y")
```

```{r}
clonality_sig_corr_df <- bind_rows(clonality_leuk_frac_corr_df, clonality_ciber_frac_corr_df, 
          clonality_mut_load_corr_df, clonality_tcr_div_corr_df,
          clonality_mrna_corr_df) %>%
    filter(p.adjust < 0.05) %>% 
    mutate(disease = factor(disease, levels = tcga_colors$Disease))
```

```{r, fig.width=5, fig.height=3}
clonality_sig_corr_df %>% 
    mutate(direction = ifelse(estimate > 0, "positive", "negative")) %>%
    group_by(disease, correlate_type, direction) %>% 
    tally() %>% 
    ungroup() %>% 
    mutate(count = ifelse(direction == "negative", -1 * n, n)) %>% 
    ggplot(aes(x = disease, y = count)) +
    geom_col(aes(fill = disease, alpha = direction), colour = "slategray") +
    # geom_violin(aes(fill = disease)) +
    # geom_quasirandom(aes(fill = disease, colour = disease), size = 0.5,
    #                  alpha = 0.3, shape = 21) +
    # ylab("miRNA-correlate correlation estimate (spearman)") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          legend.position = "top") +
    scale_fill_manual(values = tcga_colors$Color) +
    scale_alpha_manual(values = c(0.4, 1)) +
    guides(fill = FALSE) +
    facet_wrap(~ correlate_type, ncol = 2, scales = "free_y")
```

```{r, fig.width=5, fig.height=3}
clonality_mrna_corr_df %>% 
    filter((!disease %in% c("LAML", "THYM", "DLBC"))) %>%
    filter(!is.na(p.adjust),
           p.adjust < 0.05) %>%
    left_join(gene_correlate_df %>% 
                  mutate(`HGNC Symbol` =  ifelse(Gene == "TNFA",
                                                 "TNF", `HGNC Symbol`),
                         `HGNC Symbol` =  ifelse(Gene == "IL12",
                                                 "IL12A", `HGNC Symbol`),
                         `HGNC Symbol` =  ifelse(`HGNC Symbol` == "VISTA",
                                                 "C10orf54", `HGNC Symbol`)),
                  by = c("correlate" = "HGNC Symbol")) %>% 
    mutate(correlate = ifelse(is.na(`Immune Checkpoint`) & str_detect(Function, "(MHC|Granzyme)"),
                              Function, str_c("Checkpoint", `Immune Checkpoint`, "Gene")),
           correlate = ifelse(is.na(correlate), `Gene Family`, correlate)) %>% 
    select(disease, clonality, correlate, estimate, statistic, p.value, method, alternative, p.adjust, correlate_type) %>% 
    bind_rows(clonality_leuk_frac_corr_df, clonality_ciber_frac_corr_df, 
          clonality_mut_load_corr_df, clonality_tcr_div_corr_df, .) %>% 
    filter((!disease %in% c("LAML", "THYM", "DLBC"))) %>%
    filter(!is.na(p.adjust),
           p.adjust < 0.05) %>%
    mutate(correlate = fct_inorder(correlate)) %>% 
    group_by(disease, clonality, correlate, correlate_type) %>%
    summarise(estimate = mean(estimate)) %>%
    ungroup() %>%
    ggplot(aes(x = disease, y = correlate)) +
    geom_tile(aes(fill = estimate)) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

